{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: requests in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (2.31.0)\n",
      "Requirement already satisfied: crossrefapi in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (1.6.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests) (2023.11.17)\n",
      "Collecting dehyphen\n",
      "  Downloading dehyphen-0.3.4-py3-none-any.whl (18 kB)\n",
      "Collecting clean-text[gpl] (from dehyphen)\n",
      "  Downloading clean_text-0.6.0-py3-none-any.whl (11 kB)\n",
      "Collecting pd3f-flair (from dehyphen)\n",
      "  Downloading pd3f_flair-0.6.0.1-py3-none-any.whl (188 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m188.5/188.5 kB\u001b[0m \u001b[31m1.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hCollecting emoji<2.0.0,>=1.0.0 (from clean-text[gpl]->dehyphen)\n",
      "  Downloading emoji-1.7.0.tar.gz (175 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m175.4/175.4 kB\u001b[0m \u001b[31m1.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting ftfy<7.0,>=6.0 (from clean-text[gpl]->dehyphen)\n",
      "  Downloading ftfy-6.1.3-py3-none-any.whl (53 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m53.4/53.4 kB\u001b[0m \u001b[31m1.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting unidecode<2.0.0,>=1.1.1 (from clean-text[gpl]->dehyphen)\n",
      "  Downloading Unidecode-1.3.7-py3-none-any.whl (235 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m235.5/235.5 kB\u001b[0m \u001b[31m2.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting deprecated (from pd3f-flair->dehyphen)\n",
      "  Downloading Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)\n",
      "Requirement already satisfied: numpy in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from pd3f-flair->dehyphen) (1.24.2)\n",
      "Requirement already satisfied: requests in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from pd3f-flair->dehyphen) (2.31.0)\n",
      "Collecting torch (from pd3f-flair->dehyphen)\n",
      "  Downloading torch-2.1.2-cp310-none-macosx_10_9_x86_64.whl (147.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m147.0/147.0 MB\u001b[0m \u001b[31m3.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: tqdm in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from pd3f-flair->dehyphen) (4.65.0)\n",
      "Collecting wcwidth<0.3.0,>=0.2.12 (from ftfy<7.0,>=6.0->clean-text[gpl]->dehyphen)\n",
      "  Downloading wcwidth-0.2.12-py2.py3-none-any.whl (34 kB)\n",
      "Collecting wrapt<2,>=1.10 (from deprecated->pd3f-flair->dehyphen)\n",
      "  Downloading wrapt-1.16.0-cp310-cp310-macosx_10_9_x86_64.whl (37 kB)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests->pd3f-flair->dehyphen) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests->pd3f-flair->dehyphen) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests->pd3f-flair->dehyphen) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from requests->pd3f-flair->dehyphen) (2023.11.17)\n",
      "Collecting filelock (from torch->pd3f-flair->dehyphen)\n",
      "  Using cached filelock-3.13.1-py3-none-any.whl (11 kB)\n",
      "Requirement already satisfied: typing-extensions in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from torch->pd3f-flair->dehyphen) (4.5.0)\n",
      "Collecting sympy (from torch->pd3f-flair->dehyphen)\n",
      "  Using cached sympy-1.12-py3-none-any.whl (5.7 MB)\n",
      "Requirement already satisfied: networkx in /Users/fukai/projects/chromatin_reconstitution/resources/.conda/lib/python3.10/site-packages (from torch->pd3f-flair->dehyphen) (3.1)\n",
      "Collecting jinja2 (from torch->pd3f-flair->dehyphen)\n",
      "  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)\n",
      "Collecting fsspec (from torch->pd3f-flair->dehyphen)\n",
      "  Downloading fsspec-2023.12.2-py3-none-any.whl (168 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m169.0/169.0 kB\u001b[0m \u001b[31m1.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting MarkupSafe>=2.0 (from jinja2->torch->pd3f-flair->dehyphen)\n",
      "  Using cached MarkupSafe-2.1.3-cp310-cp310-macosx_10_9_x86_64.whl (13 kB)\n",
      "Collecting mpmath>=0.19 (from sympy->torch->pd3f-flair->dehyphen)\n",
      "  Using cached mpmath-1.3.0-py3-none-any.whl (536 kB)\n",
      "Building wheels for collected packages: emoji\n",
      "  Building wheel for emoji (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for emoji: filename=emoji-1.7.0-py3-none-any.whl size=171033 sha256=988f64dbf52fde41b605dfcc489e388fd961f84aaa9e2ffcb4b4a23a7e889e7c\n",
      "  Stored in directory: /Users/fukai/Library/Caches/pip/wheels/31/8a/8c/315c9e5d7773f74b33d5ed33f075b49c6eaeb7cedbb86e2cf8\n",
      "Successfully built emoji\n",
      "Installing collected packages: wcwidth, mpmath, emoji, wrapt, unidecode, sympy, MarkupSafe, ftfy, fsspec, filelock, jinja2, deprecated, clean-text, torch, pd3f-flair, dehyphen\n",
      "  Attempting uninstall: wcwidth\n",
      "    Found existing installation: wcwidth 0.2.6\n",
      "    Uninstalling wcwidth-0.2.6:\n",
      "      Successfully uninstalled wcwidth-0.2.6\n",
      "Successfully installed MarkupSafe-2.1.3 clean-text-0.6.0 dehyphen-0.3.4 deprecated-1.2.14 emoji-1.7.0 filelock-3.13.1 fsspec-2023.12.2 ftfy-6.1.3 jinja2-3.1.2 mpmath-1.3.0 pd3f-flair-0.6.0.1 sympy-1.12 torch-2.1.2 unidecode-1.3.7 wcwidth-0.2.12 wrapt-1.16.0\n"
     ]
    }
   ],
   "source": [
    "!pip install requests crossrefapi\n",
    "!pip install dehyphen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieve Crossref data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "from time import sleep\n",
    "from tqdm import tqdm\n",
    "import difflib\n",
    "import re\n",
    "from unicodedata import normalize\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_article_info_crossref(title,container_title=None):\n",
    "    base_url = \"https://api.crossref.org/works\"\n",
    "    params = {\"query.title\": title}\n",
    "    if container_title:\n",
    "        params.update({\"query.container-title\":container_title})\n",
    "    response = requests.get(base_url, params=params)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if data['message']['items']:\n",
    "            article = data['message']['items'][:3]\n",
    "            return article\n",
    "        else:\n",
    "            return {}\n",
    "    else:\n",
    "        raise RuntimeError(\"Failed to retrieve data.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_article_info_crossref2(text):\n",
    "    base_url = \"https://api.crossref.org/works\"\n",
    "    params = {\"query.bibliographic\": text}\n",
    "    response = requests.get(base_url, params=params)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        if data['message']['items']:\n",
    "            article = data['message']['items'][0]\n",
    "            return article\n",
    "        else:\n",
    "            return {}\n",
    "    else:\n",
    "        raise RuntimeError(\"Failed to retrieve data.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': '1. Barresi M J F and Gilbert S F (2019) Developmental Biology. Twelve edition, (Sinauer Associates Inc, New York).\\n',\n",
       " '2': '2. Ladoux B and Mège R-M (2017) Mechanobiology of collective cell behaviours. Nat. Rev. Mol. Cell Biol. 18: 743–757.\\n',\n",
       " '3': '3. Hollandi R, Moshkov N, Paavolainen L, Tasnadi E, Piccinini F, and Horvath P (2022) Nucleus segmentation: towards automated solutions. Trends Cell Biol. 32: 295–310.\\n',\n",
       " '4': '4. Maška M, Ulman V, Delgado-Rodriguez P, Gómez-de-Mariscal E, Nečasová T, Guerrero Peña FA, Ren TI, Meyerowitz EM, Scherr T, Löffler K, and Mikut R (2023) The cell tracking challenge: 10 years of objective benchmarking. Nat. Methods 20: 1–11.\\n',\n",
       " '5': '5. Anlaş A A and Nelson C M (2018) Tissue mechanics regulates form, function, and dysfunction. Curr. Opin. Cell Biol. 54: 98–105.\\n',\n",
       " '6': '6. Chan C J and Hiiragi T (2020) Integration of luminal pressure and signalling in tissue self-organization. Development 147: dev181297.\\n',\n",
       " '7': '7. Goodwin K and Nelson C M (2021) Mechanics of development. Dev. Cell 56: 240–250.\\n',\n",
       " '8': '8. Paci G and Mao Y (2021) Forced into shape: mechanical forces in Drosophila development and homeostasis. Semin. Cell Dev. Biol.120: 160–170.\\n',\n",
       " '9': '9. Valet M, Siggia E D, and Brivanlou A H (2022) Mechanical regulation of early vertebrate embryogenesis. Nat. Rev. Mol. Cell Biol.23: 169–184.\\n',\n",
       " '10': '10. Farahani P E and Nelson C M (2022) Revealing epithelial morphogenetic mechanisms through live imaging. Curr. Opin. Genet. Dev. 72: 61–68.\\n',\n",
       " '11': '11. Keller P J (2013) Imaging morphogenesis: technological advances and biological insights. Science 340: 1234168.\\n',\n",
       " '12': '12. Müller C B and Enderlein J (2010) Image scanning microscopy. Phys. Rev. Lett. 104: 198101.\\n',\n",
       " '13': '13. Mizuno K, Shiozawa K, Katoh TA, Minegishi K, Ide T, Ikawa Y, Nishimura H, Takaoka K, Itabashi T, Iwane AH, Nakai J (2020) Role of Ca 2+ transients at the node of the mouse embryo in breaking of left-right symmetry. Sci. Adv. 6: eaba1195.\\n',\n",
       " '14': '14. Katoh T A, Omori T, Mizuno K, Sai X, Minegishi K, Ikawa Y, Nishimura H, Itabashi T, Kajikawa E, Hiver S, and Iwane A H (2023) Immotile cilia mechanically sense the direction of fluid flow for left-right determination. Science 379: 66–71.\\n',\n",
       " '15': '15. Katoh T A, Omori T, Ishikawa T, Okada Y, and Hamada H (2023) Biophysical analysis of mechanical signals in immotile cilia of mouse embryonic nodes using advanced microscopic techniques. Bio. Protoc. 13: e4715.\\n',\n",
       " '16': '16. Shimozawa T, Yamagata K, Kondo T, Hayashi S, Shitamukai A, Konno D, Matsuzaki F, Takayama J, Onami S, Nakayama H, and Kosugi Y (2013) Improving spinning disk confocal microscopy by preventing pinhole cross-talk for intravital imaging. Proc. Natl. Acad. Sci. U.S.A 110: 3399–3404.\\n',\n",
       " '17': '17. Hayashi S and Okada Y (2015) Ultrafast superresolution fluorescence imaging with spinning disk confocal microscope optics. MBoC 26: 1743–1751.\\n',\n",
       " '18': '18. Smith A M, Mancini M C, and Nie S (2009) Second window for in vivo imaging. Nat. Nanotech. 4: 710–711.\\n',\n",
       " '19': '19. Helmchen F and Denk W (2005) Deep tissue two-photon microscopy. Nat. Methods 2: 932–940.\\n',\n",
       " '20': '20. Reynaud E G, Peychl J, Huisken J, and Tomancak P (2015) Guide to light-sheet microscopy for adventurous biologists. Nat. Methods 12: 30–34.\\n',\n",
       " '21': '21. Huisken J, Swoger J, Del Bene F, Wittbrodt J, and Stelzer E H K (2004) Optical sectioning deep inside live embryos by selective plane illumination microscopy. Science 305: 1007–1009.\\n',\n",
       " '22': '22. Bouchard M B, Voleti V, Mendes C S, Lacefield C, Grueber W B, Mann R S, Bruno R M, and Hillman E M C (2015) Swept confocally-aligned planar excitation (SCAPE) microscopy for high-speed volumetric imaging of behaving organisms. Nat. Photon. 9: 113–119.\\n',\n",
       " '23': '23. Keller P J, Schmidt A D, Wittbrodt J, and Stelzer E H K (2008) Reconstruction of zebrafish early embryonic development by scanned light sheet microscopy. Science 322: 1065–1069.\\n',\n",
       " '24': '24. Taniguchi A, Nishigami Y, Kajiura-Kobayashi H, Takao D, Tamaoki D, Nakagaki T, Nonaka S, and Sonobe S (2023) Light-sheet microscopy reveals dorsoventral asymmetric membrane dynamics of amoeba proteus during pressure-driven locomotion. Biol. Open 12: bio059671.\\n',\n",
       " '25': '25. Gao L, Shao L, Chen B-C, and Betzig E (2014) 3D Live fluo-rescence imaging of cellular dynamics using bessel beam plane illumination microscopy. Nat. Protoc. 9: 1083–101.\\n',\n",
       " '26': '26. Planchon T A, Gao L, Milkie D E, Davidson M W, Galbraith J A, Galbraith C G, and Betzig E (2011) Rapid Three-Dimensional Isotropic Imaging of Living Cells Using Bessel Beam Plane Illumination. Nat. Methods 8: 417–23.\\n',\n",
       " '27': '27. Chen B-C, Legant W R, Wang K, Shao L, Milkie D E, David-son M W, Janetopoulos C, Wu X S, Hammer III J A, Liu Z, and English B P (2014) Lattice Light-Sheet Microscopy: Imaging Molecules to Embryos at High Spatiotemporal Resolution. Science 346: 417–23.\\n',\n",
       " '28': '28. Liu T-L, Upadhyayula S, Milkie D E, Singh V, Wang K, Swinburne I A, Mosaliganti K R, Collins Z M, Hiscock T W, Shea J, Kohrman A Q (2018) Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms. Science 360: eaaq1392.\\n',\n",
       " '29': '29. McDole K, Guignard L, Amat F, Berger A, Malandain G, Royer L A, Turaga S C, Branson K, and Keller P J (2018) In Toto Imaging and Reconstruction of Post-Implantation Mouse Development at the Single-Cell Level. Cell 175: 859–876.\\n',\n",
       " '30': '30. Wang Y C, Khan Z, Kaschube M, and Wieschaus E F (2012) Differential Positioning of Adherens Junctions Is Associated with Initiation of Epithelial Folding. Nature 484: 390–393.\\n',\n",
       " '31': '31. Sun Z, Amourda C, Shagirov M, Hara Y, Saunders T E, and Toyama Y (2017) Basolateral protrusion and apical contraction cooperatively drive Drosophila germ-band extension. Nat. Cell Biol. 19: 375–383.\\n',\n",
       " '32': '32. Taniguchi K, Maeda R, Ando T, Okumura T, Nakazawa N, Hatori R, Nakamura M, Hozumi S, Fujiwara H, and Matsuno K (2011) Chirality in planar cell shape contributes to left-right asymmetric epithelial morphogenesis. Science 333: 339–341.\\n',\n",
       " '33': '33. Guirao B, Rigaud S U, Bosveld F, Bailles A, López-Gay J, Ishihara S, Sugimura K, Graner F, and Bellaïche Y (2015) Unified quantitative characterization of epithelial tissue development. eLife 4: e08519.\\n',\n",
       " '34': '34. Ishihara S and Sugimura K (2012) Bayesian inference of force dynamics during morphogenesis. J. Theor. Biol. 313: 201–211.\\n',\n",
       " '35': '35. Yamamoto T, Cockburn K, Greco V, and Kawaguchi K (2022) Probing the rules of cell coordination in live tissues by interpretable machine learning based on graph neural networks. PLoS Comput. Biol. 18: e1010477.\\n',\n",
       " '36': '36. Malin-Mayor C, Hirsch P, Guignard L, McDole K, Wan Y, Lemon W C, Kainmueller D, Keller P J, Preibisch S, and Funke J (2023) Automated reconstruction of whole-embryo cell lineages by learning from sparse annotations. Nat. Biotechnol. 41: 44–9.\\n',\n",
       " '37': '37. Wolf S, Wan Y, and McDole K (2021) Current approaches to fate mapping and lineage tracing using image data. Development 148: dev198994.\\n',\n",
       " '38': '38. de Medeiros G, Ortiz R, Strnad P, Boni A, Moos F, Repina N, Challet Meylan L, Maurer F, and Liberali P (2022) Multi-scale light-sheet organoid imaging framework. Nat. Commun.13: 4864.\\n',\n",
       " '39': '39. Suzuki T and Morishita Y (2017) A Quantitative approach to understanding vertebrate limb morphogenesis at the macroscopic tissue level. Curr. Opin. Genet. Dev. 45: 108–114.\\n',\n",
       " '40': '40. Morishita Y, Hironaka K, Lee S-W, Jin T, and Ohtsuka D (2017) Reconstructing 3D deformation dynamics for curved epithelial sheet morphogenesis from positional data of sparsely-labeled cells. Nat. Commun. 8: 15.\\n',\n",
       " '41': '41. Meijering E (2012) Cell segmentation: 50 years down the road [life sciences]. IEEE Signal Process. Mag. 29: 140–145.\\n',\n",
       " '42': '42. Xing F and Yang L (2016) Robust nucleus/cell detection and segmentation in digital pathology and microscopy images: a comprehensive review. IEEE Rev. Biomed. Eng. 9: 234–263.\\n',\n",
       " '43': '43. Nketia T A, Sailem H, Rohde G, Machiraju R, and Rittscher J (2017) Analysis of live cell images: methods, tools and opportunities. Methods 115: 65–79.\\n',\n",
       " '44': '44. Schneider C A, Rasband W S, and Eliceiri K W (2012) NIH Image to ImageJ: 25 years of image analysis. Nat. Methods 9:671–5.\\n',\n",
       " '45': '45. van der Walt S, Schönberger J L, Nunez-Iglesias J, Boulogne F, Warner J D, Yager N, Gouillart E, and Yu T (2014) Scikit-Image: image processing in python. PeerJ. 2: e453.\\n',\n",
       " '46': '46. Granger B E, and Pérez F (2021) Jupyter: Thinking and Story-telling With Code and Data. Comput. Sci. Eng. 23: 7–14.\\n',\n",
       " '47': '47. Haase R, Fazeli E, Legland D, Doube M, Culley S, Belevich I, Jokitalo E, Schorb M, Klemm A, and Tischer C (2022) A Hitchhiker’s Guide through the Bio-Image Analysis Software Universe. FEBS Let. 596: 2472–2485.\\n',\n",
       " '48': '48. Sofroniew N (2022) Napari: a multi-dimensional image viewer for Python. Zenodo. https://github.com/napari/napari/blob/main/CITATION.cff.\\n',\n",
       " '49': '49. Barbier de Reuille P, Routier-Kierzkowska A L, Kierzkowski D, Bassel G W, Schüpbach T, Tauriello G, Bajpai N, Strauss S, Weber A, Kiss A, and Burian A (2015) MorphoGraphX: a platform for quantifying morphogenesis in 4D. eLife 4: e05864.\\n',\n",
       " '50': '50. Strauss S, Runions A, Lane B, Eschweiler D, Bajpai N, Trozzi N, Routier-Kierzkowska A L, Yoshida S, da Silveira S R, Vijayan A, and Tofanelli R (2022) Using positional information to provide context for biological image analysis with MorphoGraphX 2.0. eLife 11: e72601.\\n',\n",
       " '51': '51. Aigouy B, Umetsu D, and Eaton S (2016) Segmentation and quantitative analysis of epithelial tissues. In: Dahmann C (ed.), Drosophila: Methods and Protocols, Methods in Molecular Biology, pp 227–239 (Springer, New York, NY).\\n',\n",
       " '52': '52. Aigouy B, Farhadifar R, Staple D B, Sagner A, Röper J-C, Jülicher F, and Eaton S (2010) Cell flow reorients the axis of planar polarity in the wing epithelium of Drosophila. Cell 142: 773–786.\\n',\n",
       " '53': '53. Watershed Separation. https://imagej.github.io/imaging/watershed.\\n',\n",
       " '54': '54. Zoccoler M L, Müller J R, Korten T, and Haase R (2023) Image data science with Python and Napari @EPFL. https://github.com/BiAPoL/Image-data-science-with-Python-and-Napari-EPFL2022.\\n',\n",
       " '55': '55. Stegmaier J, Amat F, Lemon W C, McDole K, Wan Y, Teodoro G, Mikut R, and Keller P J (2016) Real-time three-dimensional cell segmentation in large-scale microscopy data of developing embryos. Dev. Cell 36: 225–240.\\n',\n",
       " '56': '56. Ronneberger O, Fischer P, and Brox T (2015) U-Net: Convolutional networks for biomedical image segmentation. In: Navab N, Hornegger J, Wells W M, and Frangi A F (eds.), Med-ical Image Computing and Computer-Assisted Intervention – MICCAI 2015, pp 234–241 (Springer International Publishing, Cham).\\n',\n",
       " '57': '57. Ma J and Wang B (2023) Towards foundation models of biolog-ical image segmentation. Nat. Methods 20: 953–5.\\n',\n",
       " '58': '58. He K, Gkioxari G, Dollár P, and Girshick R (2017) Mask R-CNN. In: IEEE International Conference on Computer Vision (ICCV) 2017. Venice, Italy, pp 2961–2969.\\n',\n",
       " '59': '59. Tsai H-F, Gajda J, Sloan T F W, Rares A, and Shen A Q (2019) Usiigaci: instance-aware cell tracking in stain-free phase contrast microscopy enabled by machine learning. SoftwareX 9: 230–237.\\n',\n",
       " '60': '60. Weigert M, Schmidt U, Haase R, Sugawara K, and Myers G (2020) Star-convex polyhedra for 3D object detection and seg-mentation in microscopy. In: 2020 IEEE Winter Conference on Applications of Computer Vision (WACV) Snowmass Village, CO, USA, pp 3655–3662 (IEEE).\\n',\n",
       " '61': '61. Mandal S and Uhlmann V (2021) Splinedist: automated cell segmentation with spline curves. In: 2021 IEEE 18th Interna-tional Symposium on Biomedical Imaging (ISBI) Nice, France, pp 1082–1086.\\n',\n",
       " '62': '62. Stringer C, Wang T, Michaelos M, and Pachitariu M (2021) Cellpose: a generalist algorithm for cellular segmentation. Nat. Methods 18: 100–6.\\n',\n",
       " '63': '63. Pachitariu M and Stringer C (2022) Cellpose 2.0: how to train your own model. Nat. Methods 19: 1634–41.\\n',\n",
       " '64': '64. Lalit M, Tomancak P, and Jug F (2021) Embedding-based instance segmentation in microscopy. In: Proceedings of the Fourth Conference on Medical Imaging with Deep LearningLübeck, Germany, pp 399–415 (PMLR).\\n',\n",
       " '65': '65. DeepImageJ. https://deepimagej.github.io/.\\n',\n",
       " '66': '66. Schindelin J, Arganda-Carreras I, Frise E, Kaynig V, Longair M, Pietzsch T, Preibisch S, Rueden C, Saalfeld S, Schmid B, and Tinevez J Y (2012) Fiji: an open-source platform for biological-image analysis. Nat. Methods 9: 676–82.\\n',\n",
       " '67': '67. Gómez-de-mariscal E, García-lópez-de-haro C, Ouyang W, Donati L, Lundberg E, Unser M, Muñoz-Barrutia A, and Sage D (2021) DeepImageJ: a user-friendly environment to run deep learning models in ImageJ. Nat. Methods 18: 1192–1195.\\n',\n",
       " '68': '68. Ouyang W, Beuttenmueller F, Gómez-de-Mariscal E, Pape C, Burke T, Garcia-López-de-Haro C, Russell C, Moya-Sans L, de-la-Torre-Gutiérrez C, Schmidt D, and Kutra D (2022) BioIm-age Model Zoo: a community-driven resource for accessi-ble deep learning in bioimage analysis. bioRxiv 2022.06.07.495102.\\n',\n",
       " '69': '69. Berg S, Kutra D, Kroeger T, Straehle C N, Kausler B X, Haubold C, Schiegg M, Ales J, Beier T, Rudy M, and Eren K (2019) Ilastik: interactive machine learning for (bio)image analysis. Nat. Methods 16: 1226–1232.\\n',\n",
       " '70': '70. Ouyang W, Le T, Xu H, and Lundberg E (2021) Interactive biomedical segmentation tool powered by deep learning and ImJoy. F1000Research 10: 142.\\n',\n",
       " '71': '71. de Chaumont F, Dallongeville S, Chenouard N, Hervé N, Pop S, Provoost T, Meas-Yedid V, Pankajakshan P, Lecomte T, Le Montagner Y, and Lagache T (2012) Icy: an open bioimage informatics platform for extended reproducible research. Nat. Methods 9: 690–696.\\n',\n",
       " '72': '72. Bankhead P, Loughrey M B, Fernández J A, Dombrowski Y, McArt D G, Dunne P D, McQuaid S, Gray R T, Murray L J, Coleman H G, James J A (2017) QuPath: open source software for digital pathology image analysis. Sci. Rep. 7: 1–7.\\n',\n",
       " '73': '73. Greenwald N F, Miller G, Moen E, Kong A, Kagel A, Dougherty T, Fullaway C C, McIntosh B J, Leow K X, Schwartz M S, and Pavelchek C (2022) Whole-cell segmentation of tissue images with human-level performance using large-scale data annotation and deep learning. Nat. Biotechnol. 40: 555–565.\\n',\n",
       " '74': '74. Edlund C, Jackson T R, Khalid N, Bevan N, Dengel A, Ahmed S, Trygg J, and Sjögren R (2021) LIVECell—a large-scale dataset for label-free live cell segmentation. Nat. Methods 18: 1038–1045.\\n',\n",
       " '75': '75. Chamier L von, Laine R F, Jukkala J, Spahn C, Krentzel D, Nehme E, Lerche M, Hernández-Pérez S, Mattila P K, Karinou E, Holden S (2021) Democratising deep learning for microscopy with ZeroCostDL4Mic. Nat. Commun. 12: 2276.\\n',\n",
       " '76': '76. Masuzzo P, Troys M V, Ampe C, and Martens L (2016) Taking aim at moving targets in computational cell migration. Trends Cell Biol. 26: 88–110.\\n',\n",
       " '77': '77. Luo W, Xing J, Milan A, Zhang X, Liu W, and Kim T-K (2021) Multiple object tracking: a literature review. Artif. Intell. 293: 103448.\\n',\n",
       " '78': '78. Ciaparrone G, Luque Sánchez F, Tabik S, Troiano L, Tagliaferri R, and Herrera F (2020) Deep learning in video multi-object tracking: a survey. Neurocomputing 381: 61–88.\\n',\n",
       " '79': '79. Sugawara K, Çevrim Ç, and Averof M (2022) Tracking cell lineages in 3D by incremental deep learning. eLife 11: e69380.\\n',\n",
       " '80': '80. Amat F, Lemon W, Mossing D P, McDole K, Wan Y, Branson K, Myers E W, and Keller P J (2014) Fast, accurate reconstruction of cell lineages from large-scale fluorescence microscopy data. Nat. Methods 11: 951–8.\\n',\n",
       " '81': '81. Soelistyo C J, Ulicna K, and Lowe A R (2023) Machine learning enhanced cell tracking. Front. Bioinform. 3: 1228989.\\n',\n",
       " '82': '82. Chen Y, Song Y, Zhang C, Zhang F, O’Donnell L, Chrzanowski W, and Cai W (2021) CellTrack R-CNN: a novel end-to-end deep neural network for cell segmentation and tracking in microscopy images. In: 2021 IEEE 18th International Sympo-sium on Biomedical Imaging (ISBI) Nice, France, pp 779–782.\\n',\n",
       " '83': '83. Guignard L, Fiúza U M, Leggio B, Laussu J, Faure E, Michelin G, Biasuz K, Hufnagel L, Malandain G, Godin C, Lemaire P (2020) Contact area–dependent cell communication and the mor-phological invariance of ascidian embryogenesis. Science 369: eaar5663.\\n',\n",
       " '84': '84. Crocker J C and Grier D G (1996) Methods of digital video microscopy for colloidal studies. J. Colloid Interface Sci. 179: 298–310.\\n',\n",
       " '85': '85. Jaqaman K, Loerke D, Mettlen M, Kuwata H, Grinstein S, Schmid S L, and Danuser G (2008) Robust single-particle tracking in live-cell time-lapse sequences. Nat. Methods 5: 695-702.\\n',\n",
       " '86': '86. Tinevez J-Y, Perry N, Schindelin J, Hoopes G M, Reynolds G D, Laplantine E, Bednarek S Y, Shorte S L, and Eliceiri K W (2017) TrackMate: an open and extensible platform for single-particle tracking. Methods 115: 80–90.\\n',\n",
       " '87': '87. Ershov D, Phan M S, Pylvänäinen J W, Rigaud S U, Le Blanc L, Charles-Orszag A, Conway J R, Laine R F, Roy N H, Bonazzi D, and Duménil G (2022) TrackMate 7: integrating state-of-the-art segmentation algorithms into tracking pipelines. Nat. Methods19: 829–832.\\n',\n",
       " '88': '88. Fukai Y T and Kawaguchi K (2023) LapTrack: linear assign-ment particle tracking with tunable metrics. Bioinformatics 39: btac799.\\n',\n",
       " '89': '89. Stirling D R, Swain-Bowden M J, Lucas A M, Carpenter A E, Cimini B A, and Goodman A (2021) CellProfiler 4: improvements in speed, utility and usability. BMC Bioinf. 22: 433.\\n',\n",
       " '90': '90. Chalfoun J, Majurski M, Dima A, Halter M, Bhadriraju K, and Brady M (2016) Lineage Mapper: a versatile cell and particle tracker. Sci. Rep. 6: 36984.\\n',\n",
       " '91': '91. Ulicna K, Vallardi G, Charras G, and Lowe A R (2021) Auto-mated deep lineage tree analysis using a Bayesian single cell tracking approach. Front. Comput. Sci. 3: 734559.\\n',\n",
       " '92': '92. Schiegg M, Hanslovsky P, Kausler B X, Hufnagel L, and Hamprecht F A (2013) Conservation tracking. In: 2013 IEEE Inter-national Conference on Computer Vision Sydney, Australia, pp 2928–2935 (IEEE, ).\\n',\n",
       " '93': '93. Chenouard N, Bloch I, and Olivo-Marin J-C (2013) Multiple hypothesis tracking for cluttered biological image sequences. IEEE Trans. Pattern Anal. Mach. Intell. 35: 2736–3750.\\n',\n",
       " '94': '94. Hirsch P, Malin-Mayor C, Santella A, Preibisch S, Kainmueller D, and Funke J (2022) Tracking by weakly-supervised learning and graph optimization for whole-embryo C. elegans lineages. In: Wang L, Dou Q, Fletcher P T, Speidel S and Li S (eds.), Med-ical Image Computing and Computer Assisted Intervention – MICCAI 2022, pp 25–35 (Springer Nature Switzerland, Cham).\\n',\n",
       " '95': '95. Löffler K and Mikut R (2022) EmbedTrack—simultaneous cell segmentation and tracking through learning offsets and cluster-ing bandwidths. IEEE Access 10: 77147–77157.\\n',\n",
       " '96': '96. Ben-Haim T and Raviv T R (2022) Graph neural network for cell tracking in microscopy videos. In: Avidan S, Brostow G, Cissé M, Farinella G M, and Hassner T (eds.), Computer Vision – ECCV 2022, pp 610–626 (Springer Nature Switzerland, Cham).\\n',\n",
       " '97': '97. Moen E, Borba E, Miller G, Schwartz M, Bannon D, Koe N, Cam-plisson I, Kyme D, Pavelchek C, Price T, and Kudo T (2019) Accu-rate cell tracking and lineage construction in live-cell imaging experiments with deep learning. bioRxiv 803205.\\n',\n",
       " '98': '98. Bove A, Gradeci D, Fujita Y, Banerjee S, Charras G, and Lowe A R (2017) Local cellular neighborhood controls proliferation in cell competition. MBoC 28: 3215–3228.\\n',\n",
       " '99': '99. Ulman V, Maška M, Magnusson KE, Ronneberger O, Haubold C, Harder N, Matula P, Matula P, Svoboda D, Radojevic M, and Smal I (2017) An objective comparison of cell-tracking algorithms. Nat. Methods 14: 1141–1152.\\n',\n",
       " '100': '100. Caicedo J C, Goodman A, Karhohs KW, Cimini BA, Ackerman J, Haghighi M, Heng C, Becker T, Doan M, McQuin C, and Rohban M (2019) Nucleus segmentation across imaging experiments: the 2018 Data Science Bowl. Nat. Methods 16: 1247–1253.\\n',\n",
       " '101': '101. Nunley H, Shao B, Grover P, Singh J, Joyce B, Kim-Yip R, Kohrman A, Watters A, Gal Z, Kickuth A, Chalifoux M (2023) A novel ground truth dataset enables robust 3D nuclear instance segmentation in early mouse embryos. bioRxiv2023.03.14.532646.\\n',\n",
       " '102': '102. Maier-Hein L, Eisenmann M, Reinke A, Onogur S, Stankovic M, Scholz P, Arbel T, Bogunovic H, Bradley A P, Carass A, Feldmann C (2018) Why rankings of biomedical image analy-sis competitions should be interpreted with care. Nat. Commun.9: 5217.\\n',\n",
       " '103': '103. Chen J, Viana M P, and Rafelski S M (2023) When seeing is not believing: application-appropriate validation matters for quantitative bioimage analysis. Nat. Methods 20: 968–970.\\n',\n",
       " '104': '104. Rubens U, Mormont R, Paavolainen L, Bäcker V, Pavie B, Scholz L A, Michiels G, Maška M, Ünay D, Ball G, Hoyoux R (2020) BIAFLOWS: a collaborative framework to reproducibly deploy and benchmark bioimage analysis workflows. Patterns 1: 100040.\\n',\n",
       " '105': '105. Kirillov A, Mintun E, Ravi N, Mao H, Rolland C, Gustafson L, Xiao T, Whitehead S, Berg A C, Lo W Y, Dollár P (2023) Segment anything. arXiv 2304.02643.\\n',\n",
       " '106': '106. Arganda-Carreras I, Kaynig V, Rueden C, Eliceiri K W, Schin-delin J, Cardona A, and Sebastian Seung H (2017) Trainable Weka segmentation: a machine learning tool for microscopy pixel classification. Bioinformatics 33: 2424–2426.\\n',\n",
       " '107': '107. Arganda-Carreras I, Kaynig V, Rueden C, Schindelin J, Car-dona A, and Seung H S (2016) Trainable_Segmentation: Release v3.1.2. Zenodo. https://zenodo.org/record/59290.\\n',\n",
       " '108': '108. Arzt M, Deschamps J, Schmied C, Pietzsch T, Schmidt D, Toman-cak P, Haase R, and Jug F (2022) LABKIT: labeling and seg-mentation toolkit for big image data. Front. Comput. Sci. 4: 777728.\\n',\n",
       " '109': '109. Haase R, Pop D D, and Žigutyṫe L (2023) Haesleinhuepf/Napari-Accelerated-Pixel-and-Object-Classification: 0.12.3, Zenodo. https://zenodo.org/record/7579595.\\n',\n",
       " '110': '110. YAPiC. (2023) Yet Another Pixel Classifier (Based on Deep Learning). https://github.com/yapic/yapic. 111.\\n',\n",
       " '111': '111. Zhuang F, Qi Z, Duan K, Xi D, Zhu Y, Zhu H, Xiong H, and He Q (2021) A Comprehensive Survey on Transfer Learning. Proc. IEEE 109: 43–76.\\n',\n",
       " '112': '112. Budd S, Robinson E C, and Kainz B (2021) A survey on active learning and human-in-the-loop deep learning for medical image analysis. Med. Image Anal. 71: 102062.\\n',\n",
       " '113': '113. Wagner S, Thierbach K, Zerjatke T, Glauche I, Roeder I, and Scherf N (2021) TraCurate: Efficiently curating cell tracks. Soft-wareX 13: 100656.\\n',\n",
       " '114': '114. Cooper S, Barr A R, Glen R, and Bakal C (2017) NucliTrack: an integrated nuclei tracking application. Bioinformatics 33: 3320–3322.\\n',\n",
       " '115': '115. Padovani F, Mairhörmann B, Falter-Braun P, Lengefeld J, and Schmoller K M (2022) Segmentation, tracking and cell cycle anal-ysis of live-cell imaging data with cell-ACDC. BMC Biol. 20: 174.\\n',\n",
       " '116': '116. Han H, Wu G, Li Y, and Zi Z (2019) eDetect: a fast error detection and correction tool for live cell imaging data analysis. iScience 13: 1–8.\\n',\n",
       " '117': '117. Wait E, Winter M, Bjornsson C, Kokovay E, Wang Y, Goderie S, Temple S, and Cohen A R (2014) Visualization and correction of automated segmentation, tracking and lineaging from 5-D stem cell image sequences. BMC Bioinf. 15: 328.\\n',\n",
       " '118': '118. Downey M J, Jeziorska D M, Ott S, Tamai T K, Koentges G, Vance K W, and Bretschneider T (2011) Extracting fluorescent reporter time courses of cell lineages from high-throughput microscopy at low temporal resolution. PLOS ONE 6: e27886.\\n',\n",
       " '119': '119. Hilsenbeck O, Schwarzfischer M, Skylaki S, Schauberger B, Hoppe P S, Loeffler D, Kokkaliaris K D, Hastreiter S, Skylaki E, Filipczyk A, and Strasser M (2016) Software tools for single-cell tracking and quantification of cellular and molecular properties. Nat. Biotechnol. 34: 703–706.\\n',\n",
       " '120': '120. Piccinini F, Kiss A, and Horvath P (2016) CellTracker (not only) for dummies. Bioinformatics 32: 955–957.\\n',\n",
       " '121': '121. Cornwell J A, Li J, Mahadevan S, Draper J S, Joun G L, Zoell-ner H, Asli N S, Harvey R P, and Nordon R E (2020) Track-Pad: software for semi-automated single-cell tracking and lineage annotation. SoftwareX 11: 100440.\\n',\n",
       " '122': '122. Schott B, Traub M, Schlagenhauf C, Takamiya M, Antrit-ter T, Bartschat A, Löffler K, Blessing D, Otte J C, Kobit-ski A Y, Nienhaus G U (2018) EmbryoMiner: a new frame-work for interactive knowledge discovery in large-scale cell tracking data of developing embryos. PLoS Comput Biol. 14:e1006128.\\n',\n",
       " '123': '123. Saalfeld S, Cardona A, Hartenstein V, and Tomančák P (2009) CATMAID: collaborative annotation toolkit for massive amounts of image data. Bioinformatics 25: 1984–1986.\\n',\n",
       " '124': '124. Likar B, Maintz J B, Viergever M A, and Pernus F (2000) Ret-rospective shading correction based on entropy minimization. J. Microsc. 197: 285–295.\\n',\n",
       " '125': '125. Smith K, Li Y, Piccinini F, Csucs G, Balazs C, Bevilac-qua A, and Horvath P (2015) CIDRE: an illumination-correction method for optical microscopy. Nat. Methods12: 404–406.\\n',\n",
       " '126': '126. Peng T, Thorn K, Schroeder T, Wang L, Theis F J, Marr C, and Navab N (2017) A BaSiC tool for background and shad-ing correction of optical microscopy images. Nat. Commun. 8: 14836.\\n',\n",
       " '127': '127. Muhlich J L, Chen Y-A, Yapp C, Russell D, Santagata S, and Sorger P K (2021) Stitching and registering highly multiplexed whole slide images of tissues and tumors using ASHLAR soft-ware, preprint. Bioinformatics 38: 4613-4621.\\n',\n",
       " '128': '128. Chalfoun J, Majurski M, Blattner T, Bhadriraju K, Keyrouz W, Bajcsy P, and Brady M (2017) MIST: accurate and scalable microscopy image stitching tool with stage modeling and error minimization. Sci. Rep. 7: 4988.\\n',\n",
       " '129': '129. Fukai Y T (2021) M2Stitch. https://github.com/yfukai/m2stitch.\\n',\n",
       " '130': '130. Hörl D, Rojas Rusak F, Preusser F, Tillberg P, Randel N, Chhetri R K, Cardona A, Keller P J, Harz H, Leonhardt H, Treier M (2019) BigStitcher: reconstructing high-resolution image datasets of cleared and expanded samples. Nat. Methods 16: 870–4.\\n',\n",
       " '131': '131. Fan L, Zhang F, Fan H, and Zhang C (2019) Brief review of image denoising techniques. Vis. Comput. Ind. Biomed. Art. 2: 1–2.\\n',\n",
       " '132': '132. Ogita G, Kondo T, Ikawa K, Uemura T, Ishihara S, and Sug-imura K (2022) Image-Based Parameter Inference for Epithelial Mechanics. PLoS Comput Biol. 18: e1010209.\\n',\n",
       " '133': '133. Shindo A and Wallingford J B (2014) PCP and septins compart-mentalize cortical actomyosin to direct collective cell movement. Science 343: 649–52.\\n',\n",
       " '134': '134. Tanase M, Biais N, and Sheetz M (2007) Magnetic tweezers in cell biology. In: Methods in Cell Biology, Vol. 83, pp 473–493 (Academic Press, Cambridge, Massachusetts, USA).\\n',\n",
       " '135': '135. De Vlaminck I and Dekker C (2012) Recent advances in magnetic tweezers. Annu. Rev. Biophys. 41: 453–472.\\n',\n",
       " '136': '136. Garzon-Coral C, Fantana H A, and Howard J (2016) A force-generating machinery maintains the spindle at the cell center during mitosis. Science 352: 1124–27.\\n',\n",
       " '137': '137. Watanabe R, Okuno D, Sakakihara S, Shimabukuro K, Iino R, Yoshida M, and Noji H (2012) Mechanical modulation of catalytic power on F1-ATPase. Nat. Chem. Biol. 8: 86–92.\\n',\n",
       " '138': '138. González L M, Ruder W C, Leduc P R, and Messner W C (2014) Controlling magnetotactic bacteria through an integrated nanofabricated metallic island and optical microscope approach. Sci. Rep. 4: 86–92.\\n',\n",
       " '139': '139. Fukui H, Chow R W-Y, Xie J, Foo Y Y, Yap C H, Minc N, Mochizuki N, and Vermot J (2021) Bioelectric signaling and the control of cardiac cell identity in response to mechanical forces. Science 374: 351–354.\\n',\n",
       " '140': '140. Mongera A, Rowghanian P, Gustafson H J, Shelton E, Kealhofer D A, Carn E K, Serwane F, Lucio A A, Giammona J, and Campàs O (2018) A fluid-to-solid jamming transition underlies vertebrate body axis elongation. Nature 561(7723): 401–405.\\n',\n",
       " '141': '141. Mongera A, Pochitaloff M, Gustafson H J, Stooke-Vaughan G A, Rowghanian P, Kim S, and Campàs O (2023) Mechanics of the cellular microenvironment as probed by cells in vivo dur-ing zebrafish presomitic mesoderm differentiation. Nat. Mat. 22: 135–43.\\n',\n",
       " '142': '142. Gosse C and Croquette V (2002) Magnetic tweezers: micro-manipulation and force measurement at the molecular level. Biophys. J. 82: 3314–29.\\n',\n",
       " '143': '143. Svoboda K and Block S M (1994) Biological applica-tions of optical forces. Annu Rev Biophys Biomol Struct23: 247–85.\\n',\n",
       " '144': '144. Katoh T A, Fujimura S, and Nishizaka T (2017) 3-D single par-ticle tracking using dual images divided by prism: method and application to optical trapping. In: Ho A H-P, Kim D, and Somekh M G (eds.), Handbook of Photonics for Biomedical Engineering, pp 755–766 (Springer, Netherlands, Dordrecht).\\n',\n",
       " '145': '145. Yajima J, Mizutani K, and Nishizaka T (2008) A torque compo-nent present in mitotic kinesin Eg5 revealed by three-dimensional tracking. Nat. Struct. Mol. Biol. 15: 1119–21.\\n',\n",
       " '146': '146. Katoh T A, Ikegami K, Uchida N, Iwase T, Nakane D, Masaike T, Setou M, and Nishizaka T (2018) Three-dimensional tracking of microbeads attached to the tip of single isolated tracheal cilia beating under external load. Sci. Rep. 8: 1119–21.\\n',\n",
       " '147': '147. Naito T M, Masaike T, Nakane D, Sugawa M, Okada K A, and Nishizaka T (2019) Single-molecule pull-out manipulation of the shaft of the rotary motor F1-ATPase. Sci. Rep. 9: 7451.\\n',\n",
       " '148': '148. Bustamante C J, Chemla Y R, Liu S, and Wang M D (2021) Opti-cal tweezers in single-molecule biophysics. Nat. Rev. Methods Primers 1: 25.\\n',\n",
       " '149': '149. Mizutani M, Tulum I, Kinosita Y, Nishizaka T, and Miyata M (2018) Detailed analyses of stall force generation in mycoplasma mobile gliding. Biophys. J. 114: 1411–1419.\\n',\n",
       " '150': '150. Schonbrun E, Piestun R, Jordan P, Cooper J, Wulff K D, Courtial J, and Padgett M (2005) 3D interferometric optical tweezers using a single spatial light modulator. Opt. Express, OE 13: 3777–3786.\\n',\n",
       " '151': '151. Nishizawa K, Lin S-Z, Chardès C, Rupprecht J-F, and Lenne P-F (2023) Two-point optical manipulation reveals mechanosensitive remodeling of cell–cell contacts in vivo. Proc. Natl. Acad. Sci.120: e2212389120.\\n',\n",
       " '152': '152. Bambardekar K, Clément R, Blanc O, Chardès C, and Lenne P-F (2015) Direct laser manipulation reveals the mechanics of cell contacts in vivo. Proc. Natl. Acad. Sci. 112: 1416–1421.\\n',\n",
       " '153': '153. Djenoune L, Mahamdeh M, Truong T V, Nguyen C T, Fraser S E, Brueckner M, Howard J, and Yuan S (2023) Cilia func-tion as calcium-mediated mechanosensors that instruct left-right asymmetry. Science 379: 71–78.\\n',\n",
       " '154': '154. Krieg M, Fläschner G, Alsteens D, Gaub B M, Roos W H, Wuite G J L, Gaub H E, Gerber C, Dufrêne Y F, and Müller D J (2019) Atomic force microscopy-based mechanobiology. Nat. Rev. Phys.1: 41–57.\\n',\n",
       " '155': '155. Dufrêne Y F, Ando T, Garcia R, Alsteens D, Martinez-Martin D, Engel A, Gerber C, and Müller D J (2017) Imaging modes of atomic force microscopy for application in molecular and cell biology. Nat. Nanotech. 12: 295–307.\\n',\n",
       " '156': '156. Ando T, Uchihashi T, and Kodera N (2013) High-Speed AFM and applications to biomolecular systems. Annu. Rev. Biophys. 42: 393–414.\\n',\n",
       " '157': '157. Umeda K, McArthur S J, and Kodera N (2023) Spatiotemporal resolution in high-speed atomic force microscopy for studying biological macromolecules in action. Microscopy 72: 151–61.\\n',\n",
       " '158': '158. Chan C J, Costanzo M, Ruiz-Herrero T, Mönke G, Petrie R J, Bergert M, Diz-Muñoz A, Mahadevan L, and Hiiragi T (2019) Hydraulic control of mammalian embryo size and cell fate. Nature 571: 112–116.\\n',\n",
       " '159': '159. Shibata M, Yamashita H, Uchihashi T, Kandori H, and Ando T (2010) High-speed atomic force microscopy shows dynamic molecular processes in photoactivated bacteriorhodopsin. Nat. Nanotech. 5: 208–212.\\n',\n",
       " '160': '160. Sawyer J M, Harrell J R, Shemer G, Sullivan-Brown J, Roh-Johnson M, and Goldstein B (2010) Apical constriction: a cell shape change that can drive morphogenesis. Dev. Biol. 341: 5–19.\\n',\n",
       " '161': '161. Suzuki M, Morita H, and Ueno N (2012) Molecular mechanisms of cell shape changes that contribute to vertebrate neural tube closure. Develop. Growth Differ 54: 26—276.\\n',\n",
       " '162': '162. Martin A C and Goldstein B (2014) Apical constriction: themes and variations on a cellular mechanism driving morphogenesis. Development 141: 1987–1998.\\n',\n",
       " '163': '163. Martin A C, Kaschube M, and Wieschaus E F (2009) Pulsed con-tractions of an actin-myosin network drive apical constriction. Nature 457: 495–499.\\n',\n",
       " '164': '164. Francou A, Anderson K V, and Hadjantonakis A-K (2023) A ratchet-like apical constriction drives cell ingression during the mouse gastrulation EMT. eLife 12: e84019.\\n',\n",
       " '165': '165. Gracia M, Theis S, Proag A, Gay G, Benassayag C, and Suzanne M (2019) Mechanical impact of epithelial−mesenchymal transition on epithelial morphogenesis in Drosophila. Nat. Commun. 10: 2951.\\n',\n",
       " '166': '166. Shindo A (2018) Models of convergent extension during morpho-genesis. Wiley Interdiscip. Rev. Dev. Biol. 7: e293.\\n',\n",
       " '167': '167. Bertet C, Sulak L, and Lecuit T (2004) Myosin-dependent junc-tion remodelling controls planar cell intercalation and axis elon-gation. Nature 429: 667–671.\\n',\n",
       " '168': '168. Blankenship J T, Backovic S T, Sanny J S P, Weitz O, and Zallen J A (2006) Multicellular rosette formation links planar cell polarity to tissue morphogenesis. Dev. Cell 11: 459–470.\\n',\n",
       " '169': '169. Sui L, Alt S, Weigert M, Dye N, Eaton S, Jug F, Myers E W, Jülicher F, Salbreux G, and Dahmann C (2018) Differential lateral and basal tension drive folding of Drosophila wing discs through two distinct mechanisms. Nat. Commun. 9: 4620.\\n',\n",
       " '170': '170. John A and Rauzi M (2021) A two-tier junctional mechanism drives simultaneous tissue folding and extension. Dev. Cell 56: 1469–1483.\\n',\n",
       " '171': '171. Ohtsuka D, Kida N, Lee S-W, Kawahira N, and Mor-ishita Y (2022) Cell disorientation by loss of SHH-dependent mechanosensation causes cyclopia. Sci. Adv. 8: eabn2330.\\n',\n",
       " '172': '172. Vignes H, Vagena-Pantoula C, Prakash M, Fukui H, Norden C, Mochizuki N, Jug F, and Vermot J (2022) Extracellular mechani-cal forces drive endocardial cell volume decrease during zebrafish cardiac valve morphogenesis. Dev. Cell 57: 598–609.\\n',\n",
       " '173': '173. Okada Y, Takeda S, Tanaka Y, Belmonte J-C I, and Hirokawa N (2005) Mechanism of nodal flow: a conserved symmetry breaking event in left-right axis determination. Cell 121: 633–644.\\n',\n",
       " '174': '174. Nonaka S, Tanaka Y, Okada Y, Takeda S, Harada A, Kanai Y, Kido M, and Hirokawa N (1998) Randomization of left–right asymmetry due to loss of nodal cilia generating leftward flow of extraembryonic fluid in mice lacking KIF3B motor protein. Cell95: 829–837.\\n',\n",
       " '175': '175. McGrath J, Somlo S, Makova S, Tian X, and Brueckner M (2003) Two populations of node monocilia initiate left-right asymmetry in the mouse. Cell 114: 61–73.\\n',\n",
       " '176': '176. Yoshiba S, Shiratori H, Kuo IY, Kawasumi A, Shinohara K, Non-aka S, Asai Y, Sasaki G, Belo J A, Sasaki H, and Nakai J (2012) Cilia at the node of mouse embryos sense fluid flow for left-right determination via Pkd2. Science 338: 226–231.\\n',\n",
       " '177': '177. Nonaka S, Shiratori H, Saijoh Y, and Hamada H (2002) Determi-nation of left–right patterning of the mouse embryo by artificial nodal flow. Nature 418: 96–99.\\n',\n",
       " '178': '178. Minegishi K, Rothé B, Komatsu K R, Ono H, Ikawa Y, Nishimura H, Katoh T A, Kajikawa E, Sai X, Miyashita E, Takaoka K (2021) Fluid flow-induced left-right asymmetric decay of Dand5 mRNA in the mouse embryo requires a Bicc1-Ccr4 RNA degradation complex. Nat. Commun. 12: 4071.\\n',\n",
       " '179': '179. Supp D M, Witte D P, Potter S S, and Brueckner M (1997) Mutation of an axonemal dynein affects left–right asymmetry in inversus viscerum mice. Nature 389: 963–966.\\n',\n",
       " '180': '180. Gibbons B H and Gibbons I R (1987) Vanadate-sensitized cleav-age of dynein heavy chains by 365-Nm irradiation of demem-branated sperm flagella and its effect on the flagellar motility. J. Biol. Chem. 262: 8354–8359.\\n',\n",
       " '181': '181. Omori T, Winter K, Shinohara K, Hamada H, and Ishikawa T (2018) Simulation of the nodal flow of mutant embryos with a small number of cilia: comparison of mechanosens-ing and vesicle transport hypotheses. R. Soc. Open Sci. 5:180601.\\n',\n",
       " '182': '182. Lohoff T, Ghazanfar S, Missarova A, Koulena N, Pierson N, Grif-fiths J A, Bardot E S, Eng C H, Tyser R C, Argelaguet R, and Guibentif C. (2022) Integration of spatial and single-cell tran-scriptomic data elucidates mouse organogenesis. Nat. Biotechnol.40: 74–85.\\n',\n",
       " '183': '183. Srivatsan S R, Regier M C, Barkan E, Franks J M, Packer J S, Grosjean P, Duran M, Saxton S, Ladd J J, Spielmann M, and Lois C (2021) Embryo-scale, single-cell spatial transcriptomics. Science 373: 111–117.\\n',\n",
       " '184': '184. Goltsev Y, Samusik N, Kennedy-Darling J, Bhate S, Hale M, Vazquez G, Black S, and Nolan G P (2018) Deep profiling of mouse splenic architecture with CODEX multiplexed imaging. Cell 174: 968–981.\\n',\n",
       " '185': '185. Li X, Zhang Y, Wu J, and Dai Q (2023) Challenges and opportu-nities in Bioimage analysis. Nat. Methods 20: 958–61.\\n',\n",
       " '186': '186. Nogare D D, Hartley M, Deschamps J, Ellenberg J, and Jug F (2023) Using AI in Bioimage analysis to elevate the rate of scientific discovery as a community. Nat. Methods 20:973–5.\\n',\n",
       " '187': '187. Carpenter A E, Cimini B A, and Eliceiri K W (2023) Smart microscopes of the future. Nat. Methods 20:962–4.\\n',\n",
       " '188': '188. Moffat J, Grueneberg D A, Yang X, Kim S Y, Kloepfer A M, Hinkle G, Piqani B, Eisenhaure T M, Luo B, Grenier J K, Carpen-ter A E (2006) A lentiviral RNAi library for human and mouse genes applied to an arrayed viral high-content screen. Cell 124: 1283–1298.\\n'}"
      ]
     },
     "execution_count": 434,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"testdata.txt\",\"r\") as f:\n",
    "    original_data = f.readlines()\n",
    "original_data_with_ref_num = {}\n",
    "for d in original_data:\n",
    "    res = re.search(\"([0-9]+)\\.\",d)\n",
    "    k = res.groups(0)[0]\n",
    "    original_data_with_ref_num[k] = d\n",
    "original_data_with_ref_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'citation-number': '1.',\n",
       " 'author': [{'family': 'F', 'given': 'Barresi M.J.'},\n",
       "  {'family': 'F', 'given': 'Gilbert S.'}],\n",
       " 'title': 'Developmental Biology',\n",
       " 'edition': 'Twelve',\n",
       " 'publisher': 'Sinauer Associates Inc',\n",
       " 'language': 'pt',\n",
       " 'type': 'book',\n",
       " 'issued': {'date-parts': [[2019]]},\n",
       " 'publisher-place': 'New York'}"
      ]
     },
     "execution_count": 435,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"testdata-csl.json\") as f:\n",
    "    records = json.load(f)\n",
    "records[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 188/188 [10:00<00:00,  3.19s/it]\n"
     ]
    }
   ],
   "source": [
    "for record in tqdm(records):\n",
    "    \"\"\"\n",
    "    if \"title\" in record:\n",
    "        record[\"matched_result\"]=get_article_info_crossref(record[\"title\"],record.get('container-title',None))\n",
    "    else:\n",
    "        print(record)\n",
    "        record[\"matched_result\"]=False\n",
    "    \"\"\"\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    original_line = re.sub(\"^[0-9]\\.\",\"\",original_line)\n",
    "    record[\"matched_result\"]=get_article_info_crossref2(original_line)\n",
    "    sleep(1)\n",
    "with open(\"testdata_retrieved2.pickle\",\"wb\") as f:\n",
    "    pickle.dump(records,f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#with open(\"testdata_retrieved.pickle\",\"rb\") as f:\n",
    "#    records=pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [],
   "source": [
    "def title_normalize(s):\n",
    "    import re\n",
    "    s= s.replace(\"-\",\"\").replace(\"‐\",\"\").replace(\"’\",\"'\").replace(\" \",\" \").replace(\"–\",\"\").replace(chr(160),\" \").lower()\n",
    "    s = re.sub(\"\\.$\",\"\",s)\n",
    "    return s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "developmental biology\n",
      "scott f. gilbert—developmental biology, 2010, sinauer associates, inc., sunderland, ma ninth edition\n",
      "role of ca 2+ transients at the node of the mouse embryo in breaking of leftright symmetry\n",
      "role of ca\n",
      "            <sup>2+</sup>\n",
      "            transients at the node of the mouse embryo in breaking of leftright symmetry\n",
      "lightsheet microscopy reveals dorsoventral asymmetric membrane dynamics of amoeba proteus during pressuredriven locomotion\n",
      "lightsheet microscopy reveals dorsoventral asymmetric membrane dynamics of<i>amoeba proteus</i>during pressuredriven locomotion\n",
      "cell segmentation: 50 years down the road [life sciences\n",
      "cell segmentation: 50 years down the road [life sciences]\n",
      "napari: a multidimensional image viewer for python\n",
      "napari: a python multidimensional image viewer platform for the research community\n",
      "\n",
      "dynamic hydrologic simulation of the bear brook watershed in maine (bbwm)\n",
      "image data science with python and napari @epfl\n",
      "napari: a python multidimensional image viewer platform for the research community\n",
      "embeddingbased instance segmentation in microscopy\n",
      "embedseg: embeddingbased instance segmentation for biomedical microscopy data\n",
      "\n",
      "deepimagej: a userfriendly plugin to run deep learning models in imagej\n",
      "accurate cell tracking and lineage construction in livecell imaging experiments with deep learning\n",
      "caliban: accurate cell tracking and lineage construction in livecell imaging experiments with deep learning\n",
      "\n",
      "segment anything in glaciology: an initial study implementing the segment anything model (sam)\n",
      "trainable_segmentation: release v3.1.2\n",
      "trainable weka segmentation: a machine learning tool for microscopy pixel classification\n",
      "haesleinhuepf/napariacceleratedpixelandobjectclassification: 0.12.3\n",
      "a dictionary of indonesian prepositional constructions\n",
      "yet another pixel classifier\n",
      "yato: yet another deep learning based text analysis open toolkit\n",
      "stitching and registering highly multiplexed whole slide images of tissues and tumors using ashlar software, preprint\n",
      "stitching and registering highly multiplexed wholeslide images of tissues and tumors using ashlar\n",
      "\n",
      "introduction to github actions\n",
      "172 16\n"
     ]
    }
   ],
   "source": [
    "title_matched = []\n",
    "title_unmatched = []\n",
    "for record in records:\n",
    "    t1=title_normalize(record.get(\"title\",\"\"))\n",
    "    t2 = title_normalize(record[\"matched_result\"].get(\"title\",[\"\"])[0] if record[\"matched_result\"] else \"\")\n",
    "    if record[\"matched_result\"] and  t1 == t2:\n",
    "        title_matched.append(record.copy())\n",
    "    else:\n",
    "        print(t1)\n",
    "        print(t2)\n",
    "        title_unmatched.append(record.copy())\n",
    "print(len(title_matched),len(title_unmatched))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis for title-unmatched entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_ndiff(d):\n",
    "    res = []\n",
    "    chars = []\n",
    "    for i,s in enumerate(d):\n",
    "        if s[0]==' ': continue\n",
    "        elif s[0]=='-':\n",
    "            chars.append(s[-1])\n",
    "            res.append(u'Delete \"{}\" from position {}'.format(ord(s[-1]),i))\n",
    "        elif s[0]=='+':\n",
    "            chars.append(s[-1])\n",
    "            res.append(u'Add \"{}\" to position {}'.format(ord(s[-1]),i))   \n",
    "    return res, chars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41.\n",
      "Cell Segmentation: 50 Years Down the Road [Life Sciences]\n",
      "Cell segmentation: 50 years down the road [life sciences\n",
      "Delete \"93\" from position 56\n"
     ]
    }
   ],
   "source": [
    "for record in title_unmatched:\n",
    "    n = record[\"citation-number\"]\n",
    "    t1 = record[\"matched_result\"][\"title\"][0] if record[\"matched_result\"] else \"\"\n",
    "    t2 = record[\"title\"] if \"title\" in record else \"\"\n",
    "    res, chars = print_ndiff(difflib.ndiff(title_normalize(t1),title_normalize(t2)))\n",
    "    if len(res)>5:\n",
    "        continue\n",
    "        print(\"Too much\")\n",
    "    print(n)\n",
    "    print(t1)\n",
    "    print(t2)\n",
    "    for r in res:\n",
    "        print(r)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"title_unmatched.txt\",\"w\") as f:\n",
    "    for record in title_unmatched:\n",
    "        n = record[\"citation-number\"]\n",
    "        f.write(original_data_with_ref_num[re.sub(\"\\.$\",\"\",n)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis for title-matched citations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Authors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_name(name):\n",
    "    return name.lower()\n",
    "\n",
    "def compare_initials(initials, full_name):\n",
    "    # Split the initials and full name into parts\n",
    "    initials_parts = re.split(\"[\\-\\s\\.]\",initials)\n",
    "    full_name_parts = re.split(\"[\\-\\s]\",full_name)\n",
    "\n",
    "    # Compare each part\n",
    "    for initial, name in zip(initials_parts, full_name_parts):\n",
    "        if not name.startswith(initial.replace(\".\",\"\")):\n",
    "            return False\n",
    "\n",
    "    return True\n",
    "\n",
    "def compare_names(name1, name2):\n",
    "    # Normalize names\n",
    "    surname1, given_name1 = map(normalize_name, name1)\n",
    "    surname2, given_name2 = map(normalize_name, name2)\n",
    "\n",
    "    # Compare surnames\n",
    "    if normalize(\"NFC\",surname1).replace(\"’\",\"'\").replace(\"-\",\"\") != normalize(\"NFC\",surname2).replace(\"’\",\"'\").replace(\"-\",\"\"):\n",
    "        return False\n",
    "\n",
    "    # Compare given names\n",
    "    if len(given_name1) == len(given_name2):\n",
    "        return given_name1 == given_name2\n",
    "    elif len(given_name1) < len(given_name2):\n",
    "        return compare_initials(given_name1, given_name2)\n",
    "    else:\n",
    "        return compare_initials(given_name2, given_name1)\n",
    "\n",
    "# Test the function\n",
    "names1 = [('Ladoux', 'B'), ('Mège', 'R-M')]\n",
    "names2 = [('Ladoux', 'Benoit'), ('Mège', 'René-Marc')]\n",
    "names3 = [('Ladoux', 'Benoit'), ('Mège', 'René-Narc')]\n",
    "names4 = [('Anlaş', 'Alişya A')]\n",
    "names5 = [('Anlaş', 'A A')]\n",
    "#names6 = [('Müller', 'Claus B.'), ('Enderlein', 'Jörg')]\n",
    "#names7 = [('Müller', 'C B'), ('Enderlein', 'J')]\n",
    "\n",
    "assert all(compare_names(n1, n2) for n1, n2 in zip(names1, names2))  # Should return True\n",
    "assert not all(compare_names(n1, n2) for n1, n2 in zip(names1, names3))  # Should return False\n",
    "#assert all(compare_names(n1, n2) for n1, n2 in zip(names5, names4))  # Should return False\n",
    "#assert all(compare_names(n1, n2) for n1, n2 in zip(names7, names6))  # Should return False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_initials(s):\n",
    "    names = s.split()\n",
    "    fn = []\n",
    "    gn = []\n",
    "    for n in names:\n",
    "        n = n.strip(\".\")\n",
    "        if len(n)>1 and not all([len(nn)==1 for nn in n.split(\"-\")]):\n",
    "            fn.append(n)\n",
    "        else:\n",
    "            gn.append(n)\n",
    "    return \" \".join(fn), \" \".join(gn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Guerrero Peña FA', '') ('Guerrero Peña', 'Fidel A.')\n",
      "('Ren TI', '') ('Ren', 'Tsang Ing')\n",
      "('Meyerowitz EM', '') ('Meyerowitz', 'Elliot M.')\n",
      "('Hammer III', 'J A') ('Hammer', 'John A.')\n",
      "('Pérez', 'F') ('Perez', 'Fernando')\n",
      "('da Silveira', 'S R') ('Rodrigues da Silveira', 'Sylvia')\n",
      "('Dollár', 'P') ('Dollar', 'Piotr')\n",
      "('Dengel', 'A') ('Dale', 'Timothy')\n",
      "('Ahmed', 'S') ('Dengel', 'Andreas')\n",
      "('Trygg', 'J') ('Ahmed', 'Sheraz')\n",
      "('Sjögren', 'R') ('Trygg', 'Johan')\n",
      "('Chamier von', 'L') ('von Chamier', 'Lucas')\n",
      "('Troys', 'M V') ('Van Troys', 'Marleen')\n",
      "('Löffler', 'K') ('Loffler', 'Katharina')\n",
      "('Magnusson KE', '') ('Magnusson', 'Klas E G')\n",
      "('Karhohs KW', '') ('Karhohs', 'Kyle W.')\n",
      "('Cimini BA', '') ('Cimini', 'Beth A.')\n",
      "('Toman-cak', 'P') ('Haase', 'Robert')\n",
      "('Haase', 'R') ('Jug', 'Florian')\n",
      "('Campàs (2018) fluid-to-solid jamming transition underlies vertebrate body axis elongation Nature 561', 'O A') ('Campàs', 'Otger')\n",
      "('Kuo IY', '') ('Kuo', 'Ivana Y.')\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    # author\n",
    "    author1 = re.search(\"^[0-9]+\\. (.+)\\([0-9]\",original_line).groups()[0]\n",
    "    author1 = [s.strip() for s in re.split(\"\\,| and\",author1) if s]\n",
    "    author1 = [split_initials(a) for a in author1]\n",
    "    if \"author\" in record[\"matched_result\"]:\n",
    "        author2 = [(a[\"family\"],a[\"given\"]) for a in record['matched_result'][\"author\"] if \"family\" in a and \"given\" in a]\n",
    "    else:\n",
    "        author2 = []\n",
    "    record[\"author1\"] = author1\n",
    "    record[\"author2\"] = author2\n",
    "    if len(author1) != len(author2):\n",
    "        record[\"author_need_et_al\"] = True\n",
    "    else:\n",
    "        record[\"author_need_et_al\"] = False\n",
    "        #print(author1,author2)\n",
    "    record[\"author_valid\"] = all(compare_names(n1, n2) for n1, n2 in zip(author2, author1))\n",
    "    if not record[\"author_valid\"]:\n",
    "        for a2, a1 in zip(author2, author1):\n",
    "            if not compare_names(a2,a1):\n",
    "                print(a1,a2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4. Maška M, Ulman V, Delgado-Rodriguez P, Gómez-de-Mariscal E, Nečasová T, Guerrero Peña FA, Ren TI, Meyerowitz EM, Scherr T, Löffler K, and Mikut R (2023) The cell tracking challenge: 10 years of objective benchmarking. Nat. Methods 20: 1–11.\n",
      " >  Guerrero Peña FA  -> Guerrero Peña Fidel A.\n",
      " >  Ren TI  -> Ren Tsang Ing\n",
      " >  Meyerowitz EM  -> Meyerowitz Elliot M.\n",
      "\n",
      "27. Chen B-C, Legant W R, Wang K, Shao L, Milkie D E, David-son M W, Janetopoulos C, Wu X S, Hammer III J A, Liu Z, and English B P (2014) Lattice Light-Sheet Microscopy: Imaging Molecules to Embryos at High Spatiotemporal Resolution. Science 346: 417–23.\n",
      " >  Hammer III J A -> Hammer John A.\n",
      "\n",
      "46. Granger B E, and Pérez F (2021) Jupyter: Thinking and Story-telling With Code and Data. Comput. Sci. Eng. 23: 7–14.\n",
      " >  Pérez F -> Perez Fernando\n",
      "\n",
      "50. Strauss S, Runions A, Lane B, Eschweiler D, Bajpai N, Trozzi N, Routier-Kierzkowska A L, Yoshida S, da Silveira S R, Vijayan A, and Tofanelli R (2022) Using positional information to provide context for biological image analysis with MorphoGraphX 2.0. eLife 11: e72601.\n",
      " >  da Silveira S R -> Rodrigues da Silveira Sylvia\n",
      "\n",
      "58. He K, Gkioxari G, Dollár P, and Girshick R (2017) Mask R-CNN. In: IEEE International Conference on Computer Vision (ICCV) 2017. Venice, Italy, pp 2961–2969.\n",
      " >  Dollár P -> Dollar Piotr\n",
      "\n",
      "74. Edlund C, Jackson T R, Khalid N, Bevan N, Dengel A, Ahmed S, Trygg J, and Sjögren R (2021) LIVECell—a large-scale dataset for label-free live cell segmentation. Nat. Methods 18: 1038–1045.\n",
      " >  Dengel A -> Dale Timothy\n",
      " >  Ahmed S -> Dengel Andreas\n",
      " >  Trygg J -> Ahmed Sheraz\n",
      " >  Sjögren R -> Trygg Johan\n",
      "\n",
      "75. Chamier L von, Laine R F, Jukkala J, Spahn C, Krentzel D, Nehme E, Lerche M, Hernández-Pérez S, Mattila P K, Karinou E, Holden S (2021) Democratising deep learning for microscopy with ZeroCostDL4Mic. Nat. Commun. 12: 2276.\n",
      " >  Chamier von L -> von Chamier Lucas\n",
      "\n",
      "76. Masuzzo P, Troys M V, Ampe C, and Martens L (2016) Taking aim at moving targets in computational cell migration. Trends Cell Biol. 26: 88–110.\n",
      " >  Troys M V -> Van Troys Marleen\n",
      "\n",
      "95. Löffler K and Mikut R (2022) EmbedTrack—simultaneous cell segmentation and tracking through learning offsets and cluster-ing bandwidths. IEEE Access 10: 77147–77157.\n",
      " >  Löffler K -> Loffler Katharina\n",
      "\n",
      "99. Ulman V, Maška M, Magnusson KE, Ronneberger O, Haubold C, Harder N, Matula P, Matula P, Svoboda D, Radojevic M, and Smal I (2017) An objective comparison of cell-tracking algorithms. Nat. Methods 14: 1141–1152.\n",
      " >  Magnusson KE  -> Magnusson Klas E G\n",
      "\n",
      "100. Caicedo J C, Goodman A, Karhohs KW, Cimini BA, Ackerman J, Haghighi M, Heng C, Becker T, Doan M, McQuin C, and Rohban M (2019) Nucleus segmentation across imaging experiments: the 2018 Data Science Bowl. Nat. Methods 16: 1247–1253.\n",
      " >  Karhohs KW  -> Karhohs Kyle W.\n",
      " >  Cimini BA  -> Cimini Beth A.\n",
      "\n",
      "108. Arzt M, Deschamps J, Schmied C, Pietzsch T, Schmidt D, Toman-cak P, Haase R, and Jug F (2022) LABKIT: labeling and seg-mentation toolkit for big image data. Front. Comput. Sci. 4: 777728.\n",
      " >  Toman-cak P -> Haase Robert\n",
      " >  Haase R -> Jug Florian\n",
      "\n",
      "140. Mongera A, Rowghanian P, Gustafson H J, Shelton E, Kealhofer D A, Carn E K, Serwane F, Lucio A A, Giammona J, and Campàs O (2018) A fluid-to-solid jamming transition underlies vertebrate body axis elongation. Nature 561(7723): 401–405.\n",
      " >  Campàs (2018) fluid-to-solid jamming transition underlies vertebrate body axis elongation Nature 561 O A -> Campàs Otger\n",
      "\n",
      "176. Yoshiba S, Shiratori H, Kuo IY, Kawasumi A, Shinohara K, Non-aka S, Asai Y, Sasaki G, Belo J A, Sasaki H, and Nakai J (2012) Cilia at the node of mouse embryos sense fluid flow for left-right determination via Pkd2. Science 338: 226–231.\n",
      " >  Kuo IY  -> Kuo Ivana Y.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    if not record[\"author_valid\"]:\n",
    "        print(original_line.strip())\n",
    "        for a2, a1 in zip(record[\"author2\"],record[\"author1\"]):\n",
    "            if not compare_names(a2,a1):\n",
    "                print(\" > \", \" \".join(a1) , \"->\", \" \".join(a2))\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4. Maška M, Ulman V, Delgado-Rodriguez P, Gómez-de-Mariscal E, Nečasová T, Guerrero Peña FA, Ren TI, Meyerowitz EM, Scherr T, Löffler K, and Mikut R (2023) The cell tracking challenge: 10 years of objective benchmarking. Nat. Methods 20: 1–11.\n",
      "14. Katoh T A, Omori T, Mizuno K, Sai X, Minegishi K, Ikawa Y, Nishimura H, Itabashi T, Kajikawa E, Hiver S, and Iwane A H (2023) Immotile cilia mechanically sense the direction of fluid flow for left-right determination. Science 379: 66–71.\n",
      "16. Shimozawa T, Yamagata K, Kondo T, Hayashi S, Shitamukai A, Konno D, Matsuzaki F, Takayama J, Onami S, Nakayama H, and Kosugi Y (2013) Improving spinning disk confocal microscopy by preventing pinhole cross-talk for intravital imaging. Proc. Natl. Acad. Sci. U.S.A 110: 3399–3404.\n",
      "27. Chen B-C, Legant W R, Wang K, Shao L, Milkie D E, David-son M W, Janetopoulos C, Wu X S, Hammer III J A, Liu Z, and English B P (2014) Lattice Light-Sheet Microscopy: Imaging Molecules to Embryos at High Spatiotemporal Resolution. Science 346: 417–23.\n",
      "28. Liu T-L, Upadhyayula S, Milkie D E, Singh V, Wang K, Swinburne I A, Mosaliganti K R, Collins Z M, Hiscock T W, Shea J, Kohrman A Q (2018) Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms. Science 360: eaaq1392.\n",
      "49. Barbier de Reuille P, Routier-Kierzkowska A L, Kierzkowski D, Bassel G W, Schüpbach T, Tauriello G, Bajpai N, Strauss S, Weber A, Kiss A, and Burian A (2015) MorphoGraphX: a platform for quantifying morphogenesis in 4D. eLife 4: e05864.\n",
      "50. Strauss S, Runions A, Lane B, Eschweiler D, Bajpai N, Trozzi N, Routier-Kierzkowska A L, Yoshida S, da Silveira S R, Vijayan A, and Tofanelli R (2022) Using positional information to provide context for biological image analysis with MorphoGraphX 2.0. eLife 11: e72601.\n",
      "66. Schindelin J, Arganda-Carreras I, Frise E, Kaynig V, Longair M, Pietzsch T, Preibisch S, Rueden C, Saalfeld S, Schmid B, and Tinevez J Y (2012) Fiji: an open-source platform for biological-image analysis. Nat. Methods 9: 676–82.\n",
      "68. Ouyang W, Beuttenmueller F, Gómez-de-Mariscal E, Pape C, Burke T, Garcia-López-de-Haro C, Russell C, Moya-Sans L, de-la-Torre-Gutiérrez C, Schmidt D, and Kutra D (2022) BioIm-age Model Zoo: a community-driven resource for accessi-ble deep learning in bioimage analysis. bioRxiv 2022.06.07.495102.\n",
      "69. Berg S, Kutra D, Kroeger T, Straehle C N, Kausler B X, Haubold C, Schiegg M, Ales J, Beier T, Rudy M, and Eren K (2019) Ilastik: interactive machine learning for (bio)image analysis. Nat. Methods 16: 1226–1232.\n",
      "71. de Chaumont F, Dallongeville S, Chenouard N, Hervé N, Pop S, Provoost T, Meas-Yedid V, Pankajakshan P, Lecomte T, Le Montagner Y, and Lagache T (2012) Icy: an open bioimage informatics platform for extended reproducible research. Nat. Methods 9: 690–696.\n",
      "72. Bankhead P, Loughrey M B, Fernández J A, Dombrowski Y, McArt D G, Dunne P D, McQuaid S, Gray R T, Murray L J, Coleman H G, James J A (2017) QuPath: open source software for digital pathology image analysis. Sci. Rep. 7: 1–7.\n",
      "73. Greenwald N F, Miller G, Moen E, Kong A, Kagel A, Dougherty T, Fullaway C C, McIntosh B J, Leow K X, Schwartz M S, and Pavelchek C (2022) Whole-cell segmentation of tissue images with human-level performance using large-scale data annotation and deep learning. Nat. Biotechnol. 40: 555–565.\n",
      "74. Edlund C, Jackson T R, Khalid N, Bevan N, Dengel A, Ahmed S, Trygg J, and Sjögren R (2021) LIVECell—a large-scale dataset for label-free live cell segmentation. Nat. Methods 18: 1038–1045.\n",
      "75. Chamier L von, Laine R F, Jukkala J, Spahn C, Krentzel D, Nehme E, Lerche M, Hernández-Pérez S, Mattila P K, Karinou E, Holden S (2021) Democratising deep learning for microscopy with ZeroCostDL4Mic. Nat. Commun. 12: 2276.\n",
      "87. Ershov D, Phan M S, Pylvänäinen J W, Rigaud S U, Le Blanc L, Charles-Orszag A, Conway J R, Laine R F, Roy N H, Bonazzi D, and Duménil G (2022) TrackMate 7: integrating state-of-the-art segmentation algorithms into tracking pipelines. Nat. Methods19: 829–832.\n",
      "99. Ulman V, Maška M, Magnusson KE, Ronneberger O, Haubold C, Harder N, Matula P, Matula P, Svoboda D, Radojevic M, and Smal I (2017) An objective comparison of cell-tracking algorithms. Nat. Methods 14: 1141–1152.\n",
      "100. Caicedo J C, Goodman A, Karhohs KW, Cimini BA, Ackerman J, Haghighi M, Heng C, Becker T, Doan M, McQuin C, and Rohban M (2019) Nucleus segmentation across imaging experiments: the 2018 Data Science Bowl. Nat. Methods 16: 1247–1253.\n",
      "101. Nunley H, Shao B, Grover P, Singh J, Joyce B, Kim-Yip R, Kohrman A, Watters A, Gal Z, Kickuth A, Chalifoux M (2023) A novel ground truth dataset enables robust 3D nuclear instance segmentation in early mouse embryos. bioRxiv2023.03.14.532646.\n",
      "102. Maier-Hein L, Eisenmann M, Reinke A, Onogur S, Stankovic M, Scholz P, Arbel T, Bogunovic H, Bradley A P, Carass A, Feldmann C (2018) Why rankings of biomedical image analy-sis competitions should be interpreted with care. Nat. Commun.9: 5217.\n",
      "104. Rubens U, Mormont R, Paavolainen L, Bäcker V, Pavie B, Scholz L A, Michiels G, Maška M, Ünay D, Ball G, Hoyoux R (2020) BIAFLOWS: a collaborative framework to reproducibly deploy and benchmark bioimage analysis workflows. Patterns 1: 100040.\n",
      "108. Arzt M, Deschamps J, Schmied C, Pietzsch T, Schmidt D, Toman-cak P, Haase R, and Jug F (2022) LABKIT: labeling and seg-mentation toolkit for big image data. Front. Comput. Sci. 4: 777728.\n",
      "119. Hilsenbeck O, Schwarzfischer M, Skylaki S, Schauberger B, Hoppe P S, Loeffler D, Kokkaliaris K D, Hastreiter S, Skylaki E, Filipczyk A, and Strasser M (2016) Software tools for single-cell tracking and quantification of cellular and molecular properties. Nat. Biotechnol. 34: 703–706.\n",
      "122. Schott B, Traub M, Schlagenhauf C, Takamiya M, Antrit-ter T, Bartschat A, Löffler K, Blessing D, Otte J C, Kobit-ski A Y, Nienhaus G U (2018) EmbryoMiner: a new frame-work for interactive knowledge discovery in large-scale cell tracking data of developing embryos. PLoS Comput Biol. 14:e1006128.\n",
      "124. Likar B, Maintz J B, Viergever M A, and Pernus F (2000) Ret-rospective shading correction based on entropy minimization. J. Microsc. 197: 285–295.\n",
      "130. Hörl D, Rojas Rusak F, Preusser F, Tillberg P, Randel N, Chhetri R K, Cardona A, Keller P J, Harz H, Leonhardt H, Treier M (2019) BigStitcher: reconstructing high-resolution image datasets of cleared and expanded samples. Nat. Methods 16: 870–4.\n",
      "176. Yoshiba S, Shiratori H, Kuo IY, Kawasumi A, Shinohara K, Non-aka S, Asai Y, Sasaki G, Belo J A, Sasaki H, and Nakai J (2012) Cilia at the node of mouse embryos sense fluid flow for left-right determination via Pkd2. Science 338: 226–231.\n",
      "178. Minegishi K, Rothé B, Komatsu K R, Ono H, Ikawa Y, Nishimura H, Katoh T A, Kajikawa E, Sai X, Miyashita E, Takaoka K (2021) Fluid flow-induced left-right asymmetric decay of Dand5 mRNA in the mouse embryo requires a Bicc1-Ccr4 RNA degradation complex. Nat. Commun. 12: 4071.\n",
      "182. Lohoff T, Ghazanfar S, Missarova A, Koulena N, Pierson N, Grif-fiths J A, Bardot E S, Eng C H, Tyser R C, Argelaguet R, and Guibentif C. (2022) Integration of spatial and single-cell tran-scriptomic data elucidates mouse organogenesis. Nat. Biotechnol.40: 74–85.\n",
      "183. Srivatsan S R, Regier M C, Barkan E, Franks J M, Packer J S, Grosjean P, Duran M, Saxton S, Ladd J J, Spielmann M, and Lois C (2021) Embryo-scale, single-cell spatial transcriptomics. Science 373: 111–117.\n",
      "188. Moffat J, Grueneberg D A, Yang X, Kim S Y, Kloepfer A M, Hinkle G, Piqani B, Eisenhaure T M, Luo B, Grenier J K, Carpen-ter A E (2006) A lentiviral RNAi library for human and mouse genes applied to an arrayed viral high-content screen. Cell 124: 1283–1298.\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    if record[\"author_need_et_al\"]:\n",
    "        print(original_line.strip())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Journal etc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Journal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8. Paci G and Mao Y (2021) Forced into shape: mechanical forces in Drosophila development and homeostasis. Semin. Cell Dev. Biol.120: 160–170.\n",
      "semin. cell dev seminars cell   developmental biology\n",
      "9. Valet M, Siggia E D, and Brivanlou A H (2022) Mechanical regulation of early vertebrate embryogenesis. Nat. Rev. Mol. Cell Biol.23: 169–184.\n",
      "nat. rev. mol. cell nat rev mol cell biol\n",
      "15. Katoh T A, Omori T, Ishikawa T, Okada Y, and Hamada H (2023) Biophysical analysis of mechanical signals in immotile cilia of mouse embryonic nodes using advanced microscopic techniques. Bio. Protoc. 13: e4715.\n",
      "bio. protoc bio-protocol\n",
      "38. de Medeiros G, Ortiz R, Strnad P, Boni A, Moos F, Repina N, Challet Meylan L, Maurer F, and Liberali P (2022) Multi-scale light-sheet organoid imaging framework. Nat. Commun.13: 4864.\n",
      "nat nat commun\n",
      "51. Aigouy B, Umetsu D, and Eaton S (2016) Segmentation and quantitative analysis of epithelial tissues. In: Dahmann C (ed.), Drosophila: Methods and Protocols, Methods in Molecular Biology, pp 227–239 (Springer, New York, NY).\n",
      "drosophila: methods and protocols, methods molecular biol-ogy methods molecular biology\n",
      "56. Ronneberger O, Fischer P, and Brox T (2015) U-Net: Convolutional networks for biomedical image segmentation. In: Navab N, Hornegger J, Wells W M, and Frangi A F (eds.), Med-ical Image Computing and Computer-Assisted Intervention – MICCAI 2015, pp 234–241 (Springer International Publishing, Cham).\n",
      "med-ical image computing and computer-assisted intervention – miccai 2015 lecture notes computer science\n",
      "58. He K, Gkioxari G, Dollár P, and Girshick R (2017) Mask R-CNN. In: IEEE International Conference on Computer Vision (ICCV) 2017. Venice, Italy, pp 2961–2969.\n",
      "ieee international conference on computer vision (iccv) 2017 2017 ieee international conference on computer vision (iccv)\n",
      "61. Mandal S and Uhlmann V (2021) Splinedist: automated cell segmentation with spline curves. In: 2021 IEEE 18th Interna-tional Symposium on Biomedical Imaging (ISBI) Nice, France, pp 1082–1086.\n",
      "2021 ieee 18th interna-tional symposium on biomedical imaging (isbi 2021 ieee 18th international symposium on biomedical imaging (isbi)\n",
      "82. Chen Y, Song Y, Zhang C, Zhang F, O’Donnell L, Chrzanowski W, and Cai W (2021) CellTrack R-CNN: a novel end-to-end deep neural network for cell segmentation and tracking in microscopy images. In: 2021 IEEE 18th International Sympo-sium on Biomedical Imaging (ISBI) Nice, France, pp 779–782.\n",
      "2021 ieee 18th international sympo-sium on biomedical imaging (isbi 2021 ieee 18th international symposium on biomedical imaging (isbi)\n",
      "84. Crocker J C and Grier D G (1996) Methods of digital video microscopy for colloidal studies. J. Colloid Interface Sci. 179: 298–310.\n",
      "j. colloid interface sci journal colloid and interface science\n",
      "87. Ershov D, Phan M S, Pylvänäinen J W, Rigaud S U, Le Blanc L, Charles-Orszag A, Conway J R, Laine R F, Roy N H, Bonazzi D, and Duménil G (2022) TrackMate 7: integrating state-of-the-art segmentation algorithms into tracking pipelines. Nat. Methods19: 829–832.\n",
      "nat nat methods\n",
      "92. Schiegg M, Hanslovsky P, Kausler B X, Hufnagel L, and Hamprecht F A (2013) Conservation tracking. In: 2013 IEEE Inter-national Conference on Computer Vision Sydney, Australia, pp 2928–2935 (IEEE, ).\n",
      "2013 ieee inter-national conference on computer vision sydney, australia 2013 ieee international conference on computer vision\n",
      "94. Hirsch P, Malin-Mayor C, Santella A, Preibisch S, Kainmueller D, and Funke J (2022) Tracking by weakly-supervised learning and graph optimization for whole-embryo C. elegans lineages. In: Wang L, Dou Q, Fletcher P T, Speidel S and Li S (eds.), Med-ical Image Computing and Computer Assisted Intervention – MICCAI 2022, pp 25–35 (Springer Nature Switzerland, Cham).\n",
      "med-ical image computing and computer assisted intervention – miccai 2022 lecture notes computer science\n",
      "96. Ben-Haim T and Raviv T R (2022) Graph neural network for cell tracking in microscopy videos. In: Avidan S, Brostow G, Cissé M, Farinella G M, and Hassner T (eds.), Computer Vision – ECCV 2022, pp 610–626 (Springer Nature Switzerland, Cham).\n",
      "computer vision – eccv 2022 lecture notes computer science\n",
      "102. Maier-Hein L, Eisenmann M, Reinke A, Onogur S, Stankovic M, Scholz P, Arbel T, Bogunovic H, Bradley A P, Carass A, Feldmann C (2018) Why rankings of biomedical image analy-sis competitions should be interpreted with care. Nat. Commun.9: 5217.\n",
      "nat nat commun\n",
      "113. Wagner S, Thierbach K, Zerjatke T, Glauche I, Roeder I, and Scherf N (2021) TraCurate: Efficiently curating cell tracks. Soft-wareX 13: 100656.\n",
      "soft-warex softwarex\n",
      "125. Smith K, Li Y, Piccinini F, Csucs G, Balazs C, Bevilac-qua A, and Horvath P (2015) CIDRE: an illumination-correction method for optical microscopy. Nat. Methods12: 404–406.\n",
      "nat nat methods\n",
      "143. Svoboda K and Block S M (1994) Biological applica-tions of optical forces. Annu Rev Biophys Biomol Struct23: 247–85.\n",
      "annu rev biophys biomol annu. rev. biophys. biomol. struct.\n",
      "150. Schonbrun E, Piestun R, Jordan P, Cooper J, Wulff K D, Courtial J, and Padgett M (2005) 3D interferometric optical tweezers using a single spatial light modulator. Opt. Express, OE 13: 3777–3786.\n",
      "opt. express, oe opt. express\n",
      "151. Nishizawa K, Lin S-Z, Chardès C, Rupprecht J-F, and Lenne P-F (2023) Two-point optical manipulation reveals mechanosensitive remodeling of cell–cell contacts in vivo. Proc. Natl. Acad. Sci.120: e2212389120.\n",
      "proc. natl. acad proc. natl. acad. sci. u.s.a.\n",
      "152. Bambardekar K, Clément R, Blanc O, Chardès C, and Lenne P-F (2015) Direct laser manipulation reveals the mechanics of cell contacts in vivo. Proc. Natl. Acad. Sci. 112: 1416–1421.\n",
      "proc. natl. acad. sci proc. natl. acad. sci. u.s.a.\n",
      "154. Krieg M, Fläschner G, Alsteens D, Gaub B M, Roos W H, Wuite G J L, Gaub H E, Gerber C, Dufrêne Y F, and Müller D J (2019) Atomic force microscopy-based mechanobiology. Nat. Rev. Phys.1: 41–57.\n",
      "nat. rev nat rev phys\n",
      "166. Shindo A (2018) Models of convergent extension during morpho-genesis. Wiley Interdiscip. Rev. Dev. Biol. 7: e293.\n",
      "wiley interdiscip. rev. dev. biol wires developmental biology\n",
      "182. Lohoff T, Ghazanfar S, Missarova A, Koulena N, Pierson N, Grif-fiths J A, Bardot E S, Eng C H, Tyser R C, Argelaguet R, and Guibentif C. (2022) Integration of spatial and single-cell tran-scriptomic data elucidates mouse organogenesis. Nat. Biotechnol.40: 74–85.\n",
      "nat nat biotechnol\n"
     ]
    }
   ],
   "source": [
    "def compare_titles(j1,j2):\n",
    "    if len(j1.split()) != len(j2.split()):\n",
    "        return False\n",
    "    for s1, s2 in zip(j1.split(),j2.split()):\n",
    "        s1 = s1.replace(\".\",\"\")\n",
    "        s2 = s2.replace(\".\",\"\")\n",
    "        l = min(len(s1),len(s2))\n",
    "        if s1[:l] != s2[:l]:\n",
    "            return False\n",
    "    return True\n",
    "def _normalize_journal(s):\n",
    "    return s.replace(\"&amp;\",\" \").replace(\" in \",\" \").replace(\" of \",\" \")\n",
    "\n",
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    if 'short-container-title' in record[\"matched_result\"] and \"container-title\" in record:\n",
    "        j1 = _normalize_journal(record[\"container-title\"].lower())\n",
    "        j2 = _normalize_journal(record[\"matched_result\"]['short-container-title'][0].lower())\n",
    "        if not compare_titles(j1,j2):\n",
    "            print(original_line.strip())\n",
    "            print(j1,j2)\n",
    "        record[\"title-checked\"] = True\n",
    "    elif 'container-title' in record[\"matched_result\"] and \"container-title\" in record:\n",
    "        j1 = _normalize_journal(record[\"container-title\"].lower())\n",
    "        j2 = _normalize_journal(record[\"matched_result\"]['container-title'][0].lower())\n",
    "        if not compare_titles(j1,j2):\n",
    "            print(original_line.strip())\n",
    "            print(j1,j2)\n",
    "        record[\"title-checked\"] = True\n",
    "\n",
    "    else:\n",
    "        record[\"title-checked\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28. Liu T-L, Upadhyayula S, Milkie D E, Singh V, Wang K, Swinburne I A, Mosaliganti K R, Collins Z M, Hiscock T W, Shea J, Kohrman A Q (2018) Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms. Science 360: eaaq1392.\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    if not record[\"title-checked\"]:\n",
    "        print(original_line.strip())\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### year, volume page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 477,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'124'"
      ]
     },
     "execution_count": 477,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28. Liu T-L, Upadhyayula S, Milkie D E, Singh V, Wang K, Swinburne I A, Mosaliganti K R, Collins Z M, Hiscock T W, Shea J, Kohrman A Q (2018) Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms. Science 360: eaaq1392.\n",
      "360 / None\n",
      "87. Ershov D, Phan M S, Pylvänäinen J W, Rigaud S U, Le Blanc L, Charles-Orszag A, Conway J R, Laine R F, Roy N H, Bonazzi D, and Duménil G (2022) TrackMate 7: integrating state-of-the-art segmentation algorithms into tracking pipelines. Nat. Methods 19: 829–832.\n",
      " / 19\n",
      "88. Fukai Y T and Kawaguchi K (2023) LapTrack: linear assign-ment particle tracking with tunable metrics. Bioinformatics 39: btac799.\n",
      "39 / None\n",
      "108. Arzt M, Deschamps J, Schmied C, Pietzsch T, Schmidt D, Toman-cak P, Haase R, and Jug F (2022) LABKIT: labeling and seg-mentation toolkit for big image data. Front. Comput. Sci. 4: 777728.\n",
      "4 / None\n",
      "122. Schott B, Traub M, Schlagenhauf C, Takamiya M, Antrit-ter T, Bartschat A, Löffler K, Blessing D, Otte J C, Kobit-ski A Y, Nienhaus G U (2018) EmbryoMiner: a new frame-work for interactive knowledge discovery in large-scale cell tracking data of developing embryos. PLoS Comput Biol. 14:e1006128.\n",
      "14:e1006128 / 14\n",
      "125. Smith K, Li Y, Piccinini F, Csucs G, Balazs C, Bevilac-qua A, and Horvath P (2015) CIDRE: an illumination-correction method for optical microscopy. Nat. Methods 12: 404–406.\n",
      " / 12\n",
      "143. Svoboda K and Block S M (1994) Biological applica-tions of optical forces. Annu Rev Biophys Biomol Struct23: 247–85.\n",
      "Struct23 / 23\n",
      "154. Krieg M, Fläschner G, Alsteens D, Gaub B M, Roos W H, Wuite G J L, Gaub H E, Gerber C, Dufrêne Y F, and Müller D J (2019) Atomic force microscopy-based mechanobiology. Nat. Rev. Phys.1: 41–57.\n",
      " / 1\n",
      "174. Nonaka S, Tanaka Y, Okada Y, Takeda S, Harada A, Kanai Y, Kido M, and Hirokawa N (1998) Randomization of left–right asymmetry due to loss of nodal cilia generating leftward flow of extraembryonic fluid in mice lacking KIF3B motor protein. Cell95: 829–837.\n",
      "Cell95 / 95\n",
      "182. Lohoff T, Ghazanfar S, Missarova A, Koulena N, Pierson N, Grif-fiths J A, Bardot E S, Eng C H, Tyser R C, Argelaguet R, and Guibentif C. (2022) Integration of spatial and single-cell tran-scriptomic data elucidates mouse organogenesis. Nat. Biotechnol.40: 74–85.\n",
      " / 40\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    original_line = original_line.replace(\"Methods\",\"Methods \")\n",
    "    if not re.search(\"([0-9]+):\",original_line):\n",
    "        continue\n",
    "    r = re.search(\"[\\s\\.]([0-9]+)\\s*:\",original_line)\n",
    "    if r is None:\n",
    "        r = \"\"\n",
    "    else:\n",
    "        r = r.groups()[0]\n",
    "    v1=record.get(\"volume\",r)\n",
    "    if v1==\"\":\n",
    "        v1=r\n",
    "    if v1 is None:\n",
    "        v1=\"\"\n",
    "    else:\n",
    "        v1 = v1.split(\".\")[-1]\n",
    "        \n",
    "    if v1!=record[\"matched_result\"].get(\"volume\"):\n",
    "        print(original_line.strip())\n",
    "        print(v1,\"/\",record[\"matched_result\"].get(\"volume\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 495,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4. Maška M, Ulman V, Delgado-Rodriguez P, Gómez-de-Mariscal E, Nečasová T, Guerrero Peña FA, Ren TI, Meyerowitz EM, Scherr T, Löffler K, and Mikut R (2023) The cell tracking challenge: 10 years of objective benchmarking. Nat. Methods 20: 1–11.\n",
      "1-11 / 1010-1020\n",
      "15. Katoh T A, Omori T, Ishikawa T, Okada Y, and Hamada H (2023) Biophysical analysis of mechanical signals in immotile cilia of mouse embryonic nodes using advanced microscopic techniques. Bio. Protoc. 13: e4715.\n",
      "4715 / None\n",
      "25. Gao L, Shao L, Chen B-C, and Betzig E (2014) 3D Live fluo-rescence imaging of cellular dynamics using bessel beam plane illumination microscopy. Nat. Protoc. 9: 1083–101.\n",
      "1083-101 / 1083-1101\n",
      "26. Planchon T A, Gao L, Milkie D E, Davidson M W, Galbraith J A, Galbraith C G, and Betzig E (2011) Rapid Three-Dimensional Isotropic Imaging of Living Cells Using Bessel Beam Plane Illumination. Nat. Methods 8: 417–23.\n",
      "417-23 / 417-423\n",
      "27. Chen B-C, Legant W R, Wang K, Shao L, Milkie D E, David-son M W, Janetopoulos C, Wu X S, Hammer III J A, Liu Z, and English B P (2014) Lattice Light-Sheet Microscopy: Imaging Molecules to Embryos at High Spatiotemporal Resolution. Science 346: 417–23.\n",
      "417-23 / 1257998\n",
      "28. Liu T-L, Upadhyayula S, Milkie D E, Singh V, Wang K, Swinburne I A, Mosaliganti K R, Collins Z M, Hiscock T W, Shea J, Kohrman A Q (2018) Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms. Science 360: eaaq1392.\n",
      "1392 / None\n",
      "29. McDole K, Guignard L, Amat F, Berger A, Malandain G, Royer L A, Turaga S C, Branson K, and Keller P J (2018) In Toto Imaging and Reconstruction of Post-Implantation Mouse Development at the Single-Cell Level. Cell 175: 859–876.\n",
      "859-876 / 859-876.e33\n",
      "33. Guirao B, Rigaud S U, Bosveld F, Bailles A, López-Gay J, Ishihara S, Sugimura K, Graner F, and Bellaïche Y (2015) Unified quantitative characterization of epithelial tissue development. eLife 4: e08519.\n",
      " / None\n",
      "35. Yamamoto T, Cockburn K, Greco V, and Kawaguchi K (2022) Probing the rules of cell coordination in live tissues by interpretable machine learning based on graph neural networks. PLoS Comput. Biol. 18: e1010477.\n",
      "1010477 / e1010477\n",
      "36. Malin-Mayor C, Hirsch P, Guignard L, McDole K, Wan Y, Lemon W C, Kainmueller D, Keller P J, Preibisch S, and Funke J (2023) Automated reconstruction of whole-embryo cell lineages by learning from sparse annotations. Nat. Biotechnol. 41: 44–9.\n",
      "44-9 / 44-49\n",
      "38. de Medeiros G, Ortiz R, Strnad P, Boni A, Moos F, Repina N, Challet Meylan L, Maurer F, and Liberali P (2022) Multi-scale light-sheet organoid imaging framework. Nat. Commun.13: 4864.\n",
      "4864 / 1038/s41467-022-32465-z\n",
      "40. Morishita Y, Hironaka K, Lee S-W, Jin T, and Ohtsuka D (2017) Reconstructing 3D deformation dynamics for curved epithelial sheet morphogenesis from positional data of sparsely-labeled cells. Nat. Commun. 8: 15.\n",
      "15 / 1038/s41467-017-00023-7\n",
      "44. Schneider C A, Rasband W S, and Eliceiri K W (2012) NIH Image to ImageJ: 25 years of image analysis. Nat. Methods 9:671–5.\n",
      "671-5 / 671-675\n",
      "45. van der Walt S, Schönberger J L, Nunez-Iglesias J, Boulogne F, Warner J D, Yager N, Gouillart E, and Yu T (2014) Scikit-Image: image processing in python. PeerJ. 2: e453.\n",
      "453 / e453\n",
      "49. Barbier de Reuille P, Routier-Kierzkowska A L, Kierzkowski D, Bassel G W, Schüpbach T, Tauriello G, Bajpai N, Strauss S, Weber A, Kiss A, and Burian A (2015) MorphoGraphX: a platform for quantifying morphogenesis in 4D. eLife 4: e05864.\n",
      " / None\n",
      "50. Strauss S, Runions A, Lane B, Eschweiler D, Bajpai N, Trozzi N, Routier-Kierzkowska A L, Yoshida S, da Silveira S R, Vijayan A, and Tofanelli R (2022) Using positional information to provide context for biological image analysis with MorphoGraphX 2.0. eLife 11: e72601.\n",
      " / None\n",
      "57. Ma J and Wang B (2023) Towards foundation models of biolog-ical image segmentation. Nat. Methods 20: 953–5.\n",
      "953-5 / 953-955\n",
      "58. He K, Gkioxari G, Dollár P, and Girshick R (2017) Mask R-CNN. In: IEEE International Conference on Computer Vision (ICCV) 2017. Venice, Italy, pp 2961–2969.\n",
      "2961-2969 / None\n",
      "60. Weigert M, Schmidt U, Haase R, Sugawara K, and Myers G (2020) Star-convex polyhedra for 3D object detection and seg-mentation in microscopy. In: 2020 IEEE Winter Conference on Applications of Computer Vision (WACV) Snowmass Village, CO, USA, pp 3655–3662 (IEEE).\n",
      "3655-3662 / None\n",
      "61. Mandal S and Uhlmann V (2021) Splinedist: automated cell segmentation with spline curves. In: 2021 IEEE 18th Interna-tional Symposium on Biomedical Imaging (ISBI) Nice, France, pp 1082–1086.\n",
      "1082-1086 / None\n",
      "62. Stringer C, Wang T, Michaelos M, and Pachitariu M (2021) Cellpose: a generalist algorithm for cellular segmentation. Nat. Methods 18: 100–6.\n",
      "100-6 / 100-106\n",
      "63. Pachitariu M and Stringer C (2022) Cellpose 2.0: how to train your own model. Nat. Methods 19: 1634–41.\n",
      "1634-41 / 1634-1641\n",
      "66. Schindelin J, Arganda-Carreras I, Frise E, Kaynig V, Longair M, Pietzsch T, Preibisch S, Rueden C, Saalfeld S, Schmid B, and Tinevez J Y (2012) Fiji: an open-source platform for biological-image analysis. Nat. Methods 9: 676–82.\n",
      "676-82 / 676-682\n",
      "68. Ouyang W, Beuttenmueller F, Gómez-de-Mariscal E, Pape C, Burke T, Garcia-López-de-Haro C, Russell C, Moya-Sans L, de-la-Torre-Gutiérrez C, Schmidt D, and Kutra D (2022) BioIm-age Model Zoo: a community-driven resource for accessi-ble deep learning in bioimage analysis. bioRxiv 2022.06.07.495102.\n",
      " / None\n",
      "72. Bankhead P, Loughrey M B, Fernández J A, Dombrowski Y, McArt D G, Dunne P D, McQuaid S, Gray R T, Murray L J, Coleman H G, James J A (2017) QuPath: open source software for digital pathology image analysis. Sci. Rep. 7: 1–7.\n",
      "1-7 / None\n",
      "75. Chamier L von, Laine R F, Jukkala J, Spahn C, Krentzel D, Nehme E, Lerche M, Hernández-Pérez S, Mattila P K, Karinou E, Holden S (2021) Democratising deep learning for microscopy with ZeroCostDL4Mic. Nat. Commun. 12: 2276.\n",
      "2276 / 1038/s41467-021-22518-0\n",
      "79. Sugawara K, Çevrim Ç, and Averof M (2022) Tracking cell lineages in 3D by incremental deep learning. eLife 11: e69380.\n",
      "69380 / None\n",
      "80. Amat F, Lemon W, Mossing D P, McDole K, Wan Y, Branson K, Myers E W, and Keller P J (2014) Fast, accurate reconstruction of cell lineages from large-scale fluorescence microscopy data. Nat. Methods 11: 951–8.\n",
      "951-8 / 951-958\n",
      "81. Soelistyo C J, Ulicna K, and Lowe A R (2023) Machine learning enhanced cell tracking. Front. Bioinform. 3: 1228989.\n",
      "1228989 / None\n",
      "82. Chen Y, Song Y, Zhang C, Zhang F, O’Donnell L, Chrzanowski W, and Cai W (2021) CellTrack R-CNN: a novel end-to-end deep neural network for cell segmentation and tracking in microscopy images. In: 2021 IEEE 18th International Sympo-sium on Biomedical Imaging (ISBI) Nice, France, pp 779–782.\n",
      "779-782 / None\n",
      "83. Guignard L, Fiúza U M, Leggio B, Laussu J, Faure E, Michelin G, Biasuz K, Hufnagel L, Malandain G, Godin C, Lemaire P (2020) Contact area–dependent cell communication and the mor-phological invariance of ascidian embryogenesis. Science 369: eaar5663.\n",
      "5663 / aar5663\n",
      "88. Fukai Y T and Kawaguchi K (2023) LapTrack: linear assign-ment particle tracking with tunable metrics. Bioinformatics 39: btac799.\n",
      "799 / None\n",
      "89. Stirling D R, Swain-Bowden M J, Lucas A M, Carpenter A E, Cimini B A, and Goodman A (2021) CellProfiler 4: improvements in speed, utility and usability. BMC Bioinf. 22: 433.\n",
      "433 / None\n",
      "90. Chalfoun J, Majurski M, Dima A, Halter M, Bhadriraju K, and Brady M (2016) Lineage Mapper: a versatile cell and particle tracker. Sci. Rep. 6: 36984.\n",
      "36984 / None\n",
      "91. Ulicna K, Vallardi G, Charras G, and Lowe A R (2021) Auto-mated deep lineage tree analysis using a Bayesian single cell tracking approach. Front. Comput. Sci. 3: 734559.\n",
      "734559 / None\n",
      "92. Schiegg M, Hanslovsky P, Kausler B X, Hufnagel L, and Hamprecht F A (2013) Conservation tracking. In: 2013 IEEE Inter-national Conference on Computer Vision Sydney, Australia, pp 2928–2935 (IEEE, ).\n",
      "2928-2935 / None\n",
      "101. Nunley H, Shao B, Grover P, Singh J, Joyce B, Kim-Yip R, Kohrman A, Watters A, Gal Z, Kickuth A, Chalifoux M (2023) A novel ground truth dataset enables robust 3D nuclear instance segmentation in early mouse embryos. bioRxiv2023.03.14.532646.\n",
      " / None\n",
      "102. Maier-Hein L, Eisenmann M, Reinke A, Onogur S, Stankovic M, Scholz P, Arbel T, Bogunovic H, Bradley A P, Carass A, Feldmann C (2018) Why rankings of biomedical image analy-sis competitions should be interpreted with care. Nat. Commun.9: 5217.\n",
      " / 1038/s41467-018-07619-7\n",
      "108. Arzt M, Deschamps J, Schmied C, Pietzsch T, Schmidt D, Toman-cak P, Haase R, and Jug F (2022) LABKIT: labeling and seg-mentation toolkit for big image data. Front. Comput. Sci. 4: 777728.\n",
      "777728 / None\n",
      "115. Padovani F, Mairhörmann B, Falter-Braun P, Lengefeld J, and Schmoller K M (2022) Segmentation, tracking and cell cycle anal-ysis of live-cell imaging data with cell-ACDC. BMC Biol. 20: 174.\n",
      "174 / None\n",
      "117. Wait E, Winter M, Bjornsson C, Kokovay E, Wang Y, Goderie S, Temple S, and Cohen A R (2014) Visualization and correction of automated segmentation, tracking and lineaging from 5-D stem cell image sequences. BMC Bioinf. 15: 328.\n",
      "328 / None\n",
      "118. Downey M J, Jeziorska D M, Ott S, Tamai T K, Koentges G, Vance K W, and Bretschneider T (2011) Extracting fluorescent reporter time courses of cell lineages from high-throughput microscopy at low temporal resolution. PLOS ONE 6: e27886.\n",
      "27886 / e27886\n",
      "122. Schott B, Traub M, Schlagenhauf C, Takamiya M, Antrit-ter T, Bartschat A, Löffler K, Blessing D, Otte J C, Kobit-ski A Y, Nienhaus G U (2018) EmbryoMiner: a new frame-work for interactive knowledge discovery in large-scale cell tracking data of developing embryos. PLoS Comput Biol. 14:e1006128.\n",
      " / e1006128\n",
      "126. Peng T, Thorn K, Schroeder T, Wang L, Theis F J, Marr C, and Navab N (2017) A BaSiC tool for background and shad-ing correction of optical microscopy images. Nat. Commun. 8: 14836.\n",
      "14836 / 1038/ncomms14836\n",
      "128. Chalfoun J, Majurski M, Blattner T, Bhadriraju K, Keyrouz W, Bajcsy P, and Brady M (2017) MIST: accurate and scalable microscopy image stitching tool with stage modeling and error minimization. Sci. Rep. 7: 4988.\n",
      "4988 / None\n",
      "130. Hörl D, Rojas Rusak F, Preusser F, Tillberg P, Randel N, Chhetri R K, Cardona A, Keller P J, Harz H, Leonhardt H, Treier M (2019) BigStitcher: reconstructing high-resolution image datasets of cleared and expanded samples. Nat. Methods 16: 870–4.\n",
      "870-4 / 870-874\n",
      "131. Fan L, Zhang F, Fan H, and Zhang C (2019) Brief review of image denoising techniques. Vis. Comput. Ind. Biomed. Art. 2: 1–2.\n",
      "1-2 / None\n",
      "132. Ogita G, Kondo T, Ikawa K, Uemura T, Ishihara S, and Sug-imura K (2022) Image-Based Parameter Inference for Epithelial Mechanics. PLoS Comput Biol. 18: e1010209.\n",
      "1010209 / e1010209\n",
      "133. Shindo A and Wallingford J B (2014) PCP and septins compart-mentalize cortical actomyosin to direct collective cell movement. Science 343: 649–52.\n",
      "649-52 / 649-652\n",
      "136. Garzon-Coral C, Fantana H A, and Howard J (2016) A force-generating machinery maintains the spindle at the cell center during mitosis. Science 352: 1124–27.\n",
      "1124-27 / 1124-1127\n",
      "138. González L M, Ruder W C, Leduc P R, and Messner W C (2014) Controlling magnetotactic bacteria through an integrated nanofabricated metallic island and optical microscope approach. Sci. Rep. 4: 86–92.\n",
      "86-92 / None\n",
      "141. Mongera A, Pochitaloff M, Gustafson H J, Stooke-Vaughan G A, Rowghanian P, Kim S, and Campàs O (2023) Mechanics of the cellular microenvironment as probed by cells in vivo dur-ing zebrafish presomitic mesoderm differentiation. Nat. Mat. 22: 135–43.\n",
      "135-43 / 135-143\n",
      "142. Gosse C and Croquette V (2002) Magnetic tweezers: micro-manipulation and force measurement at the molecular level. Biophys. J. 82: 3314–29.\n",
      "3314-29 / 3314-3329\n",
      "143. Svoboda K and Block S M (1994) Biological applica-tions of optical forces. Annu Rev Biophys Biomol Struct23: 247–85.\n",
      "247-85 / 247-285\n",
      "145. Yajima J, Mizutani K, and Nishizaka T (2008) A torque compo-nent present in mitotic kinesin Eg5 revealed by three-dimensional tracking. Nat. Struct. Mol. Biol. 15: 1119–21.\n",
      "1119-21 / 1119-1121\n",
      "146. Katoh T A, Ikegami K, Uchida N, Iwase T, Nakane D, Masaike T, Setou M, and Nishizaka T (2018) Three-dimensional tracking of microbeads attached to the tip of single isolated tracheal cilia beating under external load. Sci. Rep. 8: 1119–21.\n",
      "1119-21 / None\n",
      "147. Naito T M, Masaike T, Nakane D, Sugawa M, Okada K A, and Nishizaka T (2019) Single-molecule pull-out manipulation of the shaft of the rotary motor F1-ATPase. Sci. Rep. 9: 7451.\n",
      "7451 / None\n",
      "148. Bustamante C J, Chemla Y R, Liu S, and Wang M D (2021) Opti-cal tweezers in single-molecule biophysics. Nat. Rev. Methods Primers 1: 25.\n",
      "25 / None\n",
      "150. Schonbrun E, Piestun R, Jordan P, Cooper J, Wulff K D, Courtial J, and Padgett M (2005) 3D interferometric optical tweezers using a single spatial light modulator. Opt. Express, OE 13: 3777–3786.\n",
      "3777-3786 / 3777\n",
      "151. Nishizawa K, Lin S-Z, Chardès C, Rupprecht J-F, and Lenne P-F (2023) Two-point optical manipulation reveals mechanosensitive remodeling of cell–cell contacts in vivo. Proc. Natl. Acad. Sci.120: e2212389120.\n",
      " / None\n",
      "157. Umeda K, McArthur S J, and Kodera N (2023) Spatiotemporal resolution in high-speed atomic force microscopy for studying biological macromolecules in action. Microscopy 72: 151–61.\n",
      "151-61 / 151-161\n",
      "161. Suzuki M, Morita H, and Ueno N (2012) Molecular mechanisms of cell shape changes that contribute to vertebrate neural tube closure. Develop. Growth Differ 54: 26—276.\n",
      "26-276 / 266-276\n",
      "164. Francou A, Anderson K V, and Hadjantonakis A-K (2023) A ratchet-like apical constriction drives cell ingression during the mouse gastrulation EMT. eLife 12: e84019.\n",
      "84019 / None\n",
      "165. Gracia M, Theis S, Proag A, Gay G, Benassayag C, and Suzanne M (2019) Mechanical impact of epithelial−mesenchymal transition on epithelial morphogenesis in Drosophila. Nat. Commun. 10: 2951.\n",
      "2951 / 1038/s41467-019-10720-0\n",
      "166. Shindo A (2018) Models of convergent extension during morpho-genesis. Wiley Interdiscip. Rev. Dev. Biol. 7: e293.\n",
      "293 / None\n",
      "169. Sui L, Alt S, Weigert M, Dye N, Eaton S, Jug F, Myers E W, Jülicher F, Salbreux G, and Dahmann C (2018) Differential lateral and basal tension drive folding of Drosophila wing discs through two distinct mechanisms. Nat. Commun. 9: 4620.\n",
      "4620 / 1038/s41467-018-06497-3\n",
      "170. John A and Rauzi M (2021) A two-tier junctional mechanism drives simultaneous tissue folding and extension. Dev. Cell 56: 1469–1483.\n",
      "1469-1483 / 1469-1483.e5\n",
      "171. Ohtsuka D, Kida N, Lee S-W, Kawahira N, and Mor-ishita Y (2022) Cell disorientation by loss of SHH-dependent mechanosensation causes cyclopia. Sci. Adv. 8: eabn2330.\n",
      "2330 / None\n",
      "172. Vignes H, Vagena-Pantoula C, Prakash M, Fukui H, Norden C, Mochizuki N, Jug F, and Vermot J (2022) Extracellular mechani-cal forces drive endocardial cell volume decrease during zebrafish cardiac valve morphogenesis. Dev. Cell 57: 598–609.\n",
      "598-609 / 598-609.e5\n",
      "178. Minegishi K, Rothé B, Komatsu K R, Ono H, Ikawa Y, Nishimura H, Katoh T A, Kajikawa E, Sai X, Miyashita E, Takaoka K (2021) Fluid flow-induced left-right asymmetric decay of Dand5 mRNA in the mouse embryo requires a Bicc1-Ccr4 RNA degradation complex. Nat. Commun. 12: 4071.\n",
      "4071 / 1038/s41467-021-24295-2\n",
      "181. Omori T, Winter K, Shinohara K, Hamada H, and Ishikawa T (2018) Simulation of the nodal flow of mutant embryos with a small number of cilia: comparison of mechanosens-ing and vesicle transport hypotheses. R. Soc. Open Sci. 5:180601.\n",
      " / 180601\n",
      "184. Goltsev Y, Samusik N, Kennedy-Darling J, Bhate S, Hale M, Vazquez G, Black S, and Nolan G P (2018) Deep profiling of mouse splenic architecture with CODEX multiplexed imaging. Cell 174: 968–981.\n",
      "968-981 / 968-981.e15\n",
      "185. Li X, Zhang Y, Wu J, and Dai Q (2023) Challenges and opportu-nities in Bioimage analysis. Nat. Methods 20: 958–61.\n",
      "958-61 / 958-961\n",
      "186. Nogare D D, Hartley M, Deschamps J, Ellenberg J, and Jug F (2023) Using AI in Bioimage analysis to elevate the rate of scientific discovery as a community. Nat. Methods 20:973–5.\n",
      "973-5 / 973-975\n",
      "187. Carpenter A E, Cimini B A, and Eliceiri K W (2023) Smart microscopes of the future. Nat. Methods 20:962–4.\n",
      "962-4 / 962-964\n"
     ]
    }
   ],
   "source": [
    "for j,record in enumerate(title_matched):\n",
    "    original_line = original_data_with_ref_num[re.sub(\"\\.$\",\"\",record[\"citation-number\"])]\n",
    "    p1=record.get(\"page\",\"\").replace(\"–\",\"-\")\n",
    "    p2=record[\"matched_result\"].get(\"page\")\n",
    "    if p2 is None and record[\"matched_result\"].get(\"container-title\",[\"\"])[0] in [\"Science\",\"Development\",\"Physical Review Letters\",\"Nature Communications\"]:\n",
    "        p2 = record[\"matched_result\"][\"DOI\"].split(\".\")[-1]\n",
    "    if p1 != p2:\n",
    "        print(original_line.strip())\n",
    "        print(p1,\"/\",p2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'institution': [{'name': 'bioRxiv'}],\n",
       " 'indexed': {'date-parts': [[2022, 10, 15]],\n",
       "  'date-time': '2022-10-15T10:27:08Z',\n",
       "  'timestamp': 1665829628941},\n",
       " 'posted': {'date-parts': [[2018, 1, 8]]},\n",
       " 'group-title': 'Cell Biology',\n",
       " 'reference-count': 38,\n",
       " 'publisher': 'Cold Spring Harbor Laboratory',\n",
       " 'content-domain': {'domain': [], 'crossmark-restriction': False},\n",
       " 'accepted': {'date-parts': [[2018, 1, 9]]},\n",
       " 'abstract': '<jats:title>Abstract</jats:title><jats:p>True physiological imaging of subcellular dynamics requires studying cells within their parent organisms, where all the environmental cues that drive gene expression, and hence the phenotypes we actually observe, are present. A complete understanding also requires volumetric imaging of the cell and its surroundings at high spatiotemporal resolution without inducing undue stress on either. We combined lattice light sheet microscopy with two-channel adaptive optics to achieve, across large multicellular volumes, noninvasive aberration-free imaging of subcellular processes, including endocytosis, organelle remodeling during mitosis, and the migration of axons, immune cells, and metastatic cancer cells <jats:italic>in vivo.</jats:italic> The technology reveals the phenotypic diversity within cells across different organisms and developmental stages, and may offer insights into how cells harness their intrinsic variability to adapt to different physiological environments.</jats:p><jats:sec><jats:title>One Sentence Summary</jats:title><jats:p>Combining lattice light sheet microscopy with adaptive optics enables high speed, high resolution <jats:italic>in vivo</jats:italic> 3D imaging of dynamic processes inside cells under physiological conditions within their parent organisms.</jats:p></jats:sec>',\n",
       " 'DOI': '10.1101/243352',\n",
       " 'type': 'posted-content',\n",
       " 'created': {'date-parts': [[2018, 1, 9]],\n",
       "  'date-time': '2018-01-09T06:10:38Z',\n",
       "  'timestamp': 1515478238000},\n",
       " 'source': 'Crossref',\n",
       " 'is-referenced-by-count': 5,\n",
       " 'title': ['Observing the Cell in Its Native State: Imaging Subcellular Dynamics in Multicellular Organisms'],\n",
       " 'prefix': '10.1101',\n",
       " 'author': [{'ORCID': 'http://orcid.org/0000-0001-7965-4725',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Tsung-Li',\n",
       "   'family': 'Liu',\n",
       "   'sequence': 'first',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0001-6727-5883',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Srigokul',\n",
       "   'family': 'Upadhyayula',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Daniel E.',\n",
       "   'family': 'Milkie',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Ved',\n",
       "   'family': 'Singh',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Kai',\n",
       "   'family': 'Wang',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Ian A.',\n",
       "   'family': 'Swinburne',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Kishore R.',\n",
       "   'family': 'Mosaliganti',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-8943-7967',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Zach M.',\n",
       "   'family': 'Collins',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-0319-8679',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Tom W.',\n",
       "   'family': 'Hiscock',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Jamien',\n",
       "   'family': 'Shea',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-3726-1090',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Abraham Q.',\n",
       "   'family': 'Kohrman',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Taylor N.',\n",
       "   'family': 'Medwig',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Daphne',\n",
       "   'family': 'Dambournet',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Ryan',\n",
       "   'family': 'Forster',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-6035-2134',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Brian',\n",
       "   'family': 'Cunniff',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Yuan',\n",
       "   'family': 'Ruan',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Hanako',\n",
       "   'family': 'Yashiro',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-4903-9657',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Steffen',\n",
       "   'family': 'Scholpp',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0003-4798-5153',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Elliot M.',\n",
       "   'family': 'Meyerowitz',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Dirk',\n",
       "   'family': 'Hockemeyer',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0003-3002-6271',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'David G.',\n",
       "   'family': 'Drubin',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Benjamin L.',\n",
       "   'family': 'Martin',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'David Q.',\n",
       "   'family': 'Matus',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Minoru',\n",
       "   'family': 'Koyama',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-9330-2934',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Sean G.',\n",
       "   'family': 'Megason',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'given': 'Tom',\n",
       "   'family': 'Kirchhausen',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []},\n",
       "  {'ORCID': 'http://orcid.org/0000-0002-4192-1274',\n",
       "   'authenticated-orcid': False,\n",
       "   'given': 'Eric',\n",
       "   'family': 'Betzig',\n",
       "   'sequence': 'additional',\n",
       "   'affiliation': []}],\n",
       " 'member': '246',\n",
       " 'reference': [{'key': '2019071817153877000_243352v2.1',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1126/science.1231143'},\n",
       "  {'key': '2019071817153877000_243352v2.2',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1126/science.1257998'},\n",
       "  {'key': '2019071817153877000_243352v2.3',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.4218'},\n",
       "  {'key': '2019071817153877000_243352v2.4',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1364/OE.24.024896'},\n",
       "  {'key': '2019071817153877000_243352v2.5',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1364/OE.20.013252'},\n",
       "  {'key': '2019071817153877000_243352v2.6',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1364/BOE.2.003135'},\n",
       "  {'key': '2019071817153877000_243352v2.7',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.2925'},\n",
       "  {'key': '2019071817153877000_243352v2.8',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1089/zeb.2015.1501'},\n",
       "  {'key': '2019071817153877000_243352v2.9',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1016/j.cell.2004.08.017'},\n",
       "  {'key': '2019071817153877000_243352v2.10',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1126/science.aab3500'},\n",
       "  {'key': '2019071817153877000_243352v2.11',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1083/jcb.201309096'},\n",
       "  {'key': '2019071817153877000_243352v2.12',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1091/mbc.E16-03-0164'},\n",
       "  {'key': '2019071817153877000_243352v2.13',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1083/jcb.201608071'},\n",
       "  {'key': '2019071817153877000_243352v2.14',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1091/mbc.E09-04-0327'},\n",
       "  {'key': '2019071817153877000_243352v2.15',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1002/cne.900620207'},\n",
       "  {'key': '2019071817153877000_243352v2.16',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1016/j.devcel.2014.12.001'},\n",
       "  {'key': '2019071817153877000_243352v2.17',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nrm2679'},\n",
       "  {'key': '2019071817153877000_243352v2.18',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/323712a0'},\n",
       "  {'key': '2019071817153877000_243352v2.19',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1016/0896-6273(93)90181-P'},\n",
       "  {'key': '2019071817153877000_243352v2.20',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1523/JNEUROSCI.16-03-01140.1996'},\n",
       "  {'key': '2019071817153877000_243352v2.21',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.2450'},\n",
       "  {'key': '2019071817153877000_243352v2.22',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1073/pnas.1012185108'},\n",
       "  {'key': '2019071817153877000_243352v2.23',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nri.2016.49'},\n",
       "  {'key': '2019071817153877000_243352v2.24',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nrc3628'},\n",
       "  {'key': '2019071817153877000_243352v2.25',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nrm2889'},\n",
       "  {'key': '2019071817153877000_243352v2.26',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1242/jcs.069443'},\n",
       "  {'key': '2019071817153877000_243352v2.27',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.2434'},\n",
       "  {'key': '2019071817153877000_243352v2.28',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.2062'},\n",
       "  {'key': '2019071817153877000_243352v2.29',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nmeth.3632'},\n",
       "  {'key': '2019071817153877000_243352v2.30',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1038/nprot.2014.172'},\n",
       "  {'key': '2019071817153877000_243352v2.31',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1364/OE.15.008029'},\n",
       "  {'key': '2019071817153877000_243352v2.32',\n",
       "   'doi-asserted-by': 'crossref',\n",
       "   'first-page': '094010',\n",
       "   'DOI': '10.1088/2040-8978/15/9/094010',\n",
       "   'article-title': 'Modelling of multi-conjugate adaptive optics for spatially variant aberrations in microscopy',\n",
       "   'volume': '15',\n",
       "   'year': '2013',\n",
       "   'journal-title': 'Journal of Optics'},\n",
       "  {'key': '2019071817153877000_243352v2.33',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1371/journal.pone.0134005'},\n",
       "  {'key': '2019071817153877000_243352v2.34',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1093/bioinformatics/btp184'},\n",
       "  {'key': '2019071817153877000_243352v2.35',\n",
       "   'first-page': '14',\n",
       "   'article-title': 'Distributed gradient-domain processing of planar and spherical images',\n",
       "   'volume': '29',\n",
       "   'year': '2010',\n",
       "   'journal-title': 'ACM Transactions on Graphics (TOG)'},\n",
       "  {'key': '2019071817153877000_243352v2.36',\n",
       "   'article-title': 'Gradient-domain fusion for color correction in large EM image stacks',\n",
       "   'volume': '79',\n",
       "   'year': '2015',\n",
       "   'journal-title': 'arXiv'},\n",
       "  {'key': '2019071817153877000_243352v2.37',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1371/journal.pcbi.1002780'},\n",
       "  {'key': '2019071817153877000_243352v2.38',\n",
       "   'doi-asserted-by': 'publisher',\n",
       "   'DOI': '10.1364/JOSA.70.000998'}],\n",
       " 'link': [{'URL': 'https://syndication.highwire.org/content/doi/10.1101/243352',\n",
       "   'content-type': 'unspecified',\n",
       "   'content-version': 'vor',\n",
       "   'intended-application': 'similarity-checking'}],\n",
       " 'deposited': {'date-parts': [[2020, 1, 18]],\n",
       "  'date-time': '2020-01-18T02:49:01Z',\n",
       "  'timestamp': 1579315741000},\n",
       " 'score': 111.026566,\n",
       " 'resource': {'primary': {'URL': 'http://biorxiv.org/lookup/doi/10.1101/243352'}},\n",
       " 'issued': {'date-parts': [[2018, 1, 8]]},\n",
       " 'references-count': 38,\n",
       " 'URL': 'http://dx.doi.org/10.1101/243352',\n",
       " 'relation': {'is-preprint-of': [{'id-type': 'doi',\n",
       "    'id': '10.1126/science.aaq1392',\n",
       "    'asserted-by': 'subject'}],\n",
       "  'has-review': [{'id-type': 'doi',\n",
       "    'id': '10.3410/f.732413086.793541551',\n",
       "    'asserted-by': 'object'}]},\n",
       " 'published': {'date-parts': [[2018, 1, 8]]},\n",
       " 'subtype': 'preprint'}"
      ]
     },
     "execution_count": 487,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record[\"matched_result\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
